{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0b63e5a0",
   "metadata": {},
   "source": [
    "# Crades-out cross validation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdf85b6a",
   "metadata": {},
   "source": [
    "```sh\n",
    "mkdir ../../../data/cross_validation_suppl/clade_out\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ea61e1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import random\n",
    "import os\n",
    "import math\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "matplotlib.rcParams['font.family']       = 'Arial'\n",
    "matplotlib.rcParams['font.sans-serif']   = [\"Arial\",\"DejaVu Sans\",\"Lucida Grande\",\"Verdana\"]\n",
    "matplotlib.rcParams['figure.figsize']    = [4,3]\n",
    "matplotlib.rcParams['font.size']         = 10\n",
    "matplotlib.rcParams[\"axes.labelcolor\"]   = \"#000000\"\n",
    "matplotlib.rcParams[\"axes.linewidth\"]    = 1.0\n",
    "matplotlib.rcParams[\"xtick.major.width\"] = 1.0\n",
    "matplotlib.rcParams[\"ytick.major.width\"] = 1.0\n",
    "cmap1 = plt.cm.tab20\n",
    "cmap2 = plt.cm.Set3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d90b490b",
   "metadata": {},
   "source": [
    "## Prepare clade-out validation datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0ab856a",
   "metadata": {},
   "outputs": [],
   "source": [
    "random.seed(0)\n",
    "\n",
    "# Trait table\n",
    "trait_path = \"../../../data/trait/condensed_species_NCBI.csv\"\n",
    "trait = pd.read_csv(trait_path, sep=\",\", dtype=str)\n",
    "\n",
    "# SILVA taxonomy\n",
    "silva_metadata_path = \"../../../data/silva/SILVA_138.1_SSURef_Nr99.full_metadata\"\n",
    "silva_metadata = pd.read_csv(silva_metadata_path, sep=\"\\t\", dtype=str)\n",
    "silva_metadata.fillna(\"\", inplace=True)\n",
    "\n",
    "# Concatenate trait data to SILVA based on NCBI taxid\n",
    "silva_metadata = pd.merge(trait, silva_metadata, how='inner', left_on='species_tax_id', right_on='tax_xref_ncbi', suffixes=[\"\", \"_silva\"])\n",
    "\n",
    "\n",
    "\n",
    "tax_levels = [\"order\", \"family\", \"genus\"]\n",
    "num_test_clades = 10\n",
    "num_iterations = 10\n",
    "\n",
    "for level in tax_levels:\n",
    "    dominant_clades = silva_metadata[silva_metadata[level]!=\"\"].groupby(level).size().sort_values(ascending=False).index.tolist()[:num_test_clades * num_iterations]\n",
    "    \n",
    "    dominant_clades_shuffled = random.sample(dominant_clades, len(dominant_clades))\n",
    "\n",
    "    # Split the dataset for 10-fold cross validation\n",
    "    out_dir = f'../../../data/cross_validation_suppl/clade_out/{level}'\n",
    "\n",
    "    l_tax_id = silva_metadata[\"species_tax_id\"].tolist()\n",
    "\n",
    "    ## Split\n",
    "    trait_table = pd.read_csv(\"../../../data/ref_bac2feature/trait_bac2feature.tsv\", sep=\"\\t\", dtype=str)\n",
    "    for i, test_clades in enumerate(np.array_split(dominant_clades_shuffled, num_iterations)):\n",
    "\n",
    "        silva_metadata[\"is_test_clades\"] = silva_metadata[level].apply(lambda x: x in set(test_clades))\n",
    "        test_header = silva_metadata.loc[silva_metadata[\"is_test_clades\"]][\"species_tax_id\"].tolist()\n",
    "\n",
    "        # Path\n",
    "        split_dir = os.path.join(out_dir, f\"split_{i}\")\n",
    "        if not os.path.exists(split_dir):\n",
    "            os.makedirs(split_dir)\n",
    "        test_header_path = os.path.join(split_dir, \"nodeid_test.txt\")\n",
    "        ref_header_path = os.path.join(split_dir, \"nodeid_ref.txt\")\n",
    "        ref_trait_path = os.path.join(split_dir, \"traits.tsv\")\n",
    "        # Split header\n",
    "        set_test_header = set(test_header)\n",
    "        ref_header = [h for h in l_tax_id if h not in set_test_header]\n",
    "        with open(test_header_path, 'w') as f:\n",
    "            f.write('\\n'.join(test_header))\n",
    "        with open(ref_header_path, 'w') as f:\n",
    "            f.write('\\n'.join(ref_header))\n",
    "        # Split trait table\n",
    "        is_ref = trait_table[\"species_tax_id\"].apply(\n",
    "                    lambda x: x not in set_test_header)\n",
    "        trait_table[is_ref].to_csv(ref_trait_path, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cba09f2",
   "metadata": {},
   "source": [
    "## Cross-vadalition\n",
    "```sh\n",
    "tax_levels=(\"order\" \"family\" \"genus\")\n",
    "\n",
    "for level in \"${tax_levels[@]}\"; do\n",
    "  for i in $(seq 0 9); do\n",
    "    # Split full 16S rRNA sequences\n",
    "    seqkit grep -nf ../../data/2025-05-15/$level/split_$i/nodeid_test.txt \\\n",
    "      ../../data/2024-06-06/intermediate_dir/SILVA_138.1_SSURef_NR99_tax_silva_taxid.fasta \\\n",
    "      > ../../data/2025-05-15/$level/split_$i/test_seq_full.fasta\n",
    "\n",
    "    seqkit grep -nf ../../data/2025-05-15/$level/split_$i/nodeid_ref.txt \\\n",
    "      ../../data/2024-06-06/intermediate_dir/SILVA_138.1_SSURef_NR99_tax_silva_taxid.fasta \\\n",
    "      > ../../data/2025-05-15/$level/split_$i/ref_seq_full.fasta\n",
    "  done\n",
    "\n",
    "  # Make reference dataset for PICRUSt2 using header list\n",
    "  for i in $(seq 0 9); do ./make_pro_ref.sh -t ../../../data/ref_bac2feature/phylogeny/phylogeny.tre -a ../../../data/ref_bac2feature/phylogeny/phylogeny.fasta -n ../../../data/cross_vadalition_suppl/clade_out/$level/split_$i/nodeid_ref.txt -o ../../../data/cross_vadalition_suppl/clade_out/$level/split_$i; done\n",
    "\n",
    "  # Conduct CV by Phylogenetic placement-based method\n",
    "  for i in $(seq 0 9); do bac2feature -s ../../../data/cross_vadalition_suppl/clade_out/$level/split_$i/test_seq_full.fasta -o ../../../data/cross_vadalition_suppl/clade_out/$level/split_$i/estimation_full.tsv -m phylogeny --ref_dir_placement ../../../data/cross_vadalition_suppl/clade_out/$level/split_$i --ref_trait ../../../data/cross_vadalition_suppl/clade_out/$level/split_$i/traits.tsv --threads 1 --calculate_NSTI; done\n",
    "\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff860ebb",
   "metadata": {},
   "source": [
    "## Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f6adc3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_pred_and_true_df(pred_vals_path, true_vals_path):\n",
    "    # Prediction\n",
    "    pred_vals = pd.read_csv(pred_vals_path, sep=\"\\t\")\n",
    "    pred_vals[\"sequence\"] = pred_vals[\"sequence\"].astype(str)\n",
    "    # Reference\n",
    "    true_vals = pd.read_csv(true_vals_path, sep=\"\\t\", dtype=str)\n",
    "    cmp = pd.merge(pred_vals, true_vals,\n",
    "                   left_on='sequence', right_on=\"species_tax_id\", how='inner', suffixes=['_e', '_t'])\n",
    "    return cmp\n",
    "\n",
    "def remove_null_values(cmp, t, dtype):\n",
    "    known_flag = (~cmp[t+'_t'].isnull()) & (~cmp[t+'_e'].isnull())\n",
    "    pred_vals, true_vals = cmp[known_flag][t+'_e'], cmp[known_flag][t+'_t']\n",
    "    if dtype == 'float':\n",
    "        pred_vals = pred_vals.astype(float)\n",
    "        true_vals = true_vals.astype(float)\n",
    "    elif dtype == 'int':\n",
    "        pred_vals = pred_vals.astype(int)\n",
    "        true_vals = true_vals.astype(int)\n",
    "    return pred_vals, true_vals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f35e86b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "titles = {'cell_diameter': 'Cell diameter', 'cell_length': 'Cell length', 'doubling_h': 'Doubling time', 'growth_tmp': 'Growth temp.', 'optimum_tmp': 'Optimum temp.', 'optimum_ph': 'Optimum pH', 'genome_size': 'Genome size', 'gc_content': 'GC content', 'coding_genes': 'Coding genes', 'rRNA16S_genes': 'rRNA16S genes', 'tRNA_genes': 'tRNA genes', 'gram_stain': 'Gram stain', 'sporulation': 'Sporulation', 'motility': 'Motility', 'range_salinity': 'Halophile', 'facultative_respiration': 'Facultative', 'anaerobic_respiration': 'Anaerobe', 'aerobic_respiration':'Aerobe' ,'mesophilic_range_tmp': 'Mesophile', 'thermophilic_range_tmp':'Thermophile', 'psychrophilic_range_tmp': 'Psychrophile', 'bacillus_cell_shape': 'Bacillus', 'coccus_cell_shape': 'Coccus', 'filament_cell_shape': 'Filament', 'coccobacillus_cell_shape': 'Coccobacillus', 'vibrio_cell_shape': 'Vibrio', 'spiral_cell_shape': 'Spiral'}\n",
    "\n",
    "nt = ['cell_diameter', 'cell_length', 'doubling_h', 'growth_tmp', 'optimum_tmp', 'optimum_ph', 'genome_size', 'gc_content', 'coding_genes', 'rRNA16S_genes', 'tRNA_genes']\n",
    "\n",
    "ct = ['gram_stain',\n",
    "      'sporulation', 'motility', 'range_salinity', 'facultative_respiration',\n",
    "      'anaerobic_respiration', 'aerobic_respiration', 'mesophilic_range_tmp',\n",
    "      'thermophilic_range_tmp', 'psychrophilic_range_tmp',\n",
    "      'bacillus_cell_shape', 'coccus_cell_shape', 'filament_cell_shape',\n",
    "      'coccobacillus_cell_shape', 'vibrio_cell_shape', 'spiral_cell_shape']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e30df22a",
   "metadata": {},
   "outputs": [],
   "source": [
    "true_vals_path = \"../../../data/ref_bac2feature/trait_bac2feature.tsv\"\n",
    "true_vals = pd.read_csv(true_vals_path, sep=\"\\t\", dtype=str)\n",
    "\n",
    "for i in range(0, 10):\n",
    "    # Genus\n",
    "    level = \"genus\"\n",
    "    genus_result_path = f\"../../../data/cross_validation_suppl/clade_out/{level}/split_{i}/estimation_full.tsv\"\n",
    "    cmp = get_pred_and_true_df(pred_vals_path=genus_result_path, true_vals_path=true_vals_path)\n",
    "    res_list_genus = []\n",
    "    for t in nt:\n",
    "        pred_vals, true_vals = remove_null_values(cmp=cmp, t=t, dtype='float')\n",
    "        res_list_genus += [pred_vals.corr(true_vals, method=\"pearson\")]\n",
    "\n",
    "    # Family\n",
    "    level = \"family\"\n",
    "    family_result_path = f\"../../../data/cross_validation_suppl/clade_out/{level}/split_{i}/estimation_full.tsv\"\n",
    "    cmp = get_pred_and_true_df(pred_vals_path=family_result_path, true_vals_path=true_vals_path)\n",
    "    res_list_family = [] \n",
    "    for t in nt:\n",
    "        pred_vals, true_vals = remove_null_values(cmp=cmp, t=t, dtype='float')\n",
    "        res_list_family += [pred_vals.corr(true_vals, method=\"pearson\")]\n",
    "\n",
    "    # Order\n",
    "    level = \"order\"\n",
    "    order_result_path = f\"../../../data/cross_validation_suppl/clade_out/{level}/split_{i}/estimation_full.tsv\"\n",
    "    cmp = get_pred_and_true_df(pred_vals_path=order_result_path, true_vals_path=true_vals_path)\n",
    "    res_list_order = []\n",
    "    for t in nt:\n",
    "        pred_vals, true_vals = remove_null_values(cmp=cmp, t=t, dtype='float')\n",
    "        res_list_order += [pred_vals.corr(true_vals, method=\"pearson\")]\n",
    "\n",
    "    # 10-fold cross validation\n",
    "    pic_result_path = f\"../../../data/cross_validation/split_{i}/estimation_full.tsv\"\n",
    "    cmp = get_pred_and_true_df(pred_vals_path=pic_result_path, true_vals_path=true_vals_path)\n",
    "    res_list_pic = []\n",
    "    for t in nt:\n",
    "        pred_vals, true_vals = remove_null_values(cmp=cmp, t=t, dtype='float')\n",
    "        res_list_pic += [pred_vals.corr(true_vals, method=\"pearson\")]\n",
    "\n",
    "    # Concat\n",
    "    tmp = pd.DataFrame({\"trait\": nt,\n",
    "                        \"Genus\": res_list_genus,\n",
    "                        \"Family\": res_list_family,\n",
    "                        \"Order\": res_list_order,\n",
    "                        \"Random\": res_list_pic\n",
    "                        })\n",
    "    tmp[\"split\"] = i\n",
    "    if i == 0:\n",
    "        res_all_nt = tmp\n",
    "    else:\n",
    "        res_all_nt = pd.concat([res_all_nt, tmp], axis=0)\n",
    "\n",
    "res_all_melt_nt = res_all_nt.melt(id_vars=['trait', 'split'], var_name='method', value_name='accuracy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e7e8b264",
   "metadata": {},
   "outputs": [],
   "source": [
    "ntraits = len(nt)\n",
    "ncols = 4\n",
    "nrows = math.ceil(float(ntraits) / ncols)\n",
    "fig, axes = plt.subplots(nrows, ncols, figsize=(1.5*ncols, 1.5*nrows))\n",
    "\n",
    "\n",
    "for i, t in enumerate(nt):\n",
    "    ax = axes.flatten()[i]\n",
    "    sns.boxplot(data=res_all_melt_nt[res_all_melt_nt[\"trait\"] == t],\n",
    "                x=\"method\", y=\"accuracy\", hue=\"method\", ax=ax, order=[\"Order\", \"Family\", \"Genus\", \"Random\"],\n",
    "                palette=\"Set2\", fill=False, linewidth=1.3, showfliers=False)\n",
    "    sns.stripplot(data=res_all_melt_nt[res_all_melt_nt[\"trait\"] == t],\n",
    "                  x=\"method\", y=\"accuracy\",\n",
    "                  ax=ax, order=[\"Order\", \"Family\", \"Genus\", \"Random\"], color=\"black\", size=3, jitter=0.2)\n",
    "    ax.axhline(0.5, color='gray', linestyle='--', linewidth=0.5)\n",
    "    ax.set_ylim(-0.1, 1)\n",
    "    ax.set_title(titles[t], fontsize=10)\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.set_xticks(list(range(len([\"Order\", \"Family\", \"Genus\", \"Random\"]))))\n",
    "    if i // ncols == nrows - 1:\n",
    "        ax.set_xticklabels([\"Order\", \"Family\", \"Genus\", \"10-fold\"], rotation=90)\n",
    "    else:\n",
    "        ax.set_xticklabels([])\n",
    "    if i % ncols == 0:\n",
    "        ax.set_ylabel(\"Accuracy\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "\n",
    "axes.flatten()[-1].axis('off')\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"../../../results/09_cross_validation_suppl/figS7a.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "571c2b28",
   "metadata": {},
   "outputs": [],
   "source": [
    "true_vals_path = \"../../../data/ref_bac2feature/trait_bac2feature.tsv\"\n",
    "true_vals = pd.read_csv(true_vals_path, sep=\"\\t\", dtype=str)\n",
    "\n",
    "for i in range(0, 10):\n",
    "    # Genus\n",
    "    level = \"genus\"\n",
    "    genus_result_path = f\"../../../data/cross_validation_suppl/clade_out/{level}/split_{i}/estimation_full.tsv\"\n",
    "    cmp = get_pred_and_true_df(pred_vals_path=genus_result_path, true_vals_path=true_vals_path)\n",
    "    res_list_genus = []\n",
    "    for t in ct:\n",
    "        pred_vals, true_vals = remove_null_values(cmp=cmp, t=t, dtype='float')\n",
    "        res_list_genus += [pred_vals.corr(true_vals, method=\"pearson\")]\n",
    "\n",
    "    # Family\n",
    "    level = \"family\"\n",
    "    family_result_path = f\"../../../data/cross_validation_suppl/clade_out/{level}/split_{i}/estimation_full.tsv\"\n",
    "    cmp = get_pred_and_true_df(pred_vals_path=family_result_path, true_vals_path=true_vals_path)\n",
    "    res_list_family = [] \n",
    "    for t in ct:\n",
    "        pred_vals, true_vals = remove_null_values(cmp=cmp, t=t, dtype='float')\n",
    "        res_list_family += [pred_vals.corr(true_vals, method=\"pearson\")]\n",
    "\n",
    "    # Order\n",
    "    level = \"order\"\n",
    "    order_result_path = f\"../../../data/cross_validation_suppl/clade_out/{level}/split_{i}/estimation_full.tsv\"\n",
    "    cmp = get_pred_and_true_df(pred_vals_path=order_result_path, true_vals_path=true_vals_path)\n",
    "    res_list_order = []\n",
    "    for t in ct:\n",
    "        pred_vals, true_vals = remove_null_values(cmp=cmp, t=t, dtype='float')\n",
    "        res_list_order += [pred_vals.corr(true_vals, method=\"pearson\")]\n",
    "\n",
    "    # 10-fold cross validation\n",
    "    pic_result_path = f\"../../../data/cross_validation/split_{i}/estimation_full.tsv\"\n",
    "    cmp = get_pred_and_true_df(pred_vals_path=pic_result_path, true_vals_path=true_vals_path)\n",
    "    res_list_pic = []\n",
    "    for t in ct:\n",
    "        pred_vals, true_vals = remove_null_values(cmp=cmp, t=t, dtype='float')\n",
    "        res_list_pic += [pred_vals.corr(true_vals, method=\"pearson\")]\n",
    "\n",
    "    # Concat\n",
    "    tmp = pd.DataFrame({\"trait\": ct,\n",
    "                        \"Genus\": res_list_genus,\n",
    "                        \"Family\": res_list_family,\n",
    "                        \"Order\": res_list_order,\n",
    "                        \"Random\": res_list_pic\n",
    "                        })\n",
    "    tmp[\"split\"] = i\n",
    "    if i == 0:\n",
    "        res_all_ct = tmp\n",
    "    else:\n",
    "        res_all_ct = pd.concat([res_all_ct, tmp], axis=0)\n",
    "\n",
    "res_all_melt_ct = res_all_ct.melt(id_vars=['trait', 'split'], var_name='method', value_name='accuracy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "185d5d84",
   "metadata": {},
   "outputs": [],
   "source": [
    "ntraits = len(ct)\n",
    "ncols = 4\n",
    "nrows = math.ceil(float(ntraits) / ncols)\n",
    "fig, axes = plt.subplots(nrows, ncols, figsize=(1.5*ncols, 1.5*nrows))\n",
    "\n",
    "\n",
    "for i, t in enumerate(ct):\n",
    "    ax = axes.flatten()[i]\n",
    "    # sns.boxplot(x=\"method\", y=\"accuracy\", data=res_all_melt_nt[res_all_melt_nt[\"trait\"] == t],\n",
    "    #             ax=ax,linewidth=0.5)\n",
    "    sns.boxplot(data=res_all_melt_ct[res_all_melt_ct[\"trait\"] == t],\n",
    "                x=\"method\", y=\"accuracy\", hue=\"method\", ax=ax, order=[\"Order\", \"Family\", \"Genus\", \"Random\"],\n",
    "                palette=\"Set2\", fill=False, linewidth=1.3, showfliers=False)\n",
    "    sns.stripplot(data=res_all_melt_ct[res_all_melt_ct[\"trait\"] == t],\n",
    "                  x=\"method\", y=\"accuracy\", order=[\"Order\", \"Family\", \"Genus\", \"Random\"],\n",
    "                  ax=ax, color=\"black\", size=3, jitter=0.2)\n",
    "    ax.axhline(0.5, color='gray', linestyle='--', linewidth=0.5)\n",
    "    ax.set_ylim(-0.1, 1)\n",
    "    ax.set_title(titles[t], fontsize=10)\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.set_xticks(list(range(len([\"Order\", \"Family\", \"Genus\", \"Random\"]))))\n",
    "    if i // ncols == nrows - 1:\n",
    "        ax.set_xticklabels([\"Order\", \"Family\", \"Genus\", \"10-fold\"], rotation=90)\n",
    "    else:\n",
    "        ax.set_xticklabels([])\n",
    "    if i % ncols == 0:\n",
    "        ax.set_ylabel(\"Accuracy\")\n",
    "    else:\n",
    "        ax.set_ylabel(\"\")\n",
    "\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"../../../results/09_cross_validation_suppl/figS7b.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bac2feature",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
