{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocess trait and 16S rRNA gene sequence data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment settings\n",
    "```sh\n",
    "# Working Directory\n",
    "cd Bac2fFeature/scripts/02_data_preprocessing\n",
    "# Conda environment\n",
    "conda activate bac2feature_experiment\n",
    "# Output Directory\n",
    "directories=(\n",
    "    \"../../data/\"\n",
    "    \"../../data/silva\"\n",
    "    \"../../data/trait\"\n",
    "    \"../../data/intermediate_dir\"\n",
    "    \"../../data/ref_bac2feature\"\n",
    "    \"../../data/ref_bac2feature/homology\"\n",
    "    \"../../data/ref_bac2feature/taxonomy\"\n",
    "    \"../../data/ref_bac2feature/phylogeny\"\n",
    ")\n",
    "for dir in \"${directories[@]}\"; do\n",
    "  if [ ! -d \"$dir\" ]; then\n",
    "    mkdir -p \"$dir\"\n",
    "  fi\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import math\n",
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "from collections import Counter\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio import Phylo, SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Concatenattion of the trait data to 16S rRNA gene sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path\n",
    "out_dir = '../../data/'\n",
    "silva_dir = os.path.join(out_dir, 'silva')\n",
    "trait_dir = os.path.join(out_dir, 'trait')\n",
    "intermediate_dir = os.path.join(out_dir, 'intermediate_dir')\n",
    "\n",
    "# Trait data by Madin et al., 2020\n",
    "madin_path = os.path.join(trait_dir, 'condensed_species_NCBI.csv')\n",
    "madin_table = pd.read_csv(madin_path, sep=',', dtype='object')\n",
    "\n",
    "# SILVA SSU Ref NR99\n",
    "silva_metadata_path = os.path.join(silva_dir, 'SILVA_138.1_SSURef_Nr99.full_metadata')\n",
    "silva_metadata = pd.read_csv(silva_metadata_path, sep='\\t', dtype='object')\n",
    "silva_metadata['accession'] = silva_metadata[\"acc\"] + \".\" + silva_metadata[\"start\"] + \".\" + silva_metadata[\"stop\"]\n",
    "\n",
    "# Concatenate trait data to SILVA based on NCBI taxid\n",
    "madin_silva = pd.merge(madin_table, silva_metadata, how='inner', left_on='species_tax_id', right_on='tax_xref_ncbi', suffixes=[\"\", \"_silva\"])\n",
    "\n",
    "# Select a sequence with the hightest align_quality_slv and align_bp_score_slv\n",
    "madin_silva.sort_values(by=[\"align_quality_slv\", \"align_bp_score_slv\"], ascending=False, inplace=True)\n",
    "madin_silva.drop_duplicates(subset='species_tax_id', keep='first', inplace=True)\n",
    "madin_silva.reset_index(drop=True, inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preprocessing the trait data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Rename ambiguous trait names\n",
    "ms_format = madin_silva.fillna(\"NA\")\n",
    "ms_format = ms_format.rename(columns={'d1_lo': 'cell_diameter', 'd2_lo': 'cell_length', 'metabolism': 'respiration'})\n",
    "\n",
    "tax_cols = ['species', 'genus', 'family', 'order', 'class', 'phylum', 'superkingdom']\n",
    "nt_cols = ['cell_diameter', 'cell_length', 'doubling_h', 'growth_tmp', 'optimum_tmp', 'optimum_ph', 'genome_size', 'gc_content', 'coding_genes', 'rRNA16S_genes', 'tRNA_genes']\n",
    "ct_cols = ['gram_stain', 'respiration', 'sporulation', 'motility', 'range_tmp', 'range_salinity', 'cell_shape']\n",
    "\n",
    "# Cell size and doubling time: take log-10 scale\n",
    "def calc_log(x):\n",
    "    return math.log10(float(x)) if x != \"NA\" else \"NA\"\n",
    "for c in ['cell_diameter', 'cell_length', 'doubling_h']:\n",
    "    ms_format[c] = ms_format[c].apply(lambda x: calc_log(x))\n",
    "\n",
    "# Categorical traits: reorganize caterogies and remove minor categories\n",
    "rules = {\"respiration\": {\"obligate aerobic\": \"aerobic\",\n",
    "                         \"obligate anaerobic\": \"anaerobic\",\n",
    "                         \"microaerophilic\": \"aerobic\"},\n",
    "         \"motility\": {\"flagella\": \"yes\",\n",
    "                      \"gliding\": \"yes\",\n",
    "                      \"axial filament\": \"yes\"},\n",
    "         \"range_tmp\": {\"psychrotolerant\": \"psychrophilic\",\n",
    "                       \"extreme thermophilic\": \"thermophilic\",\n",
    "                       \"thermotolerant\": \"thermophilic\",\n",
    "                       \"facultative psychrophilic\": \"psychrophilic\"},\n",
    "         \"range_salinity\": {\"moderate-halophilic\": \"halophilic\",\n",
    "                            \"halotolerant\": \"halophilic\",\n",
    "                            \"extreme-halophilic\": \"halophilic\"}}\n",
    "for t, rule in rules.items():\n",
    "    for k, v in rule.items():\n",
    "        ms_format[t] = ms_format[t].replace(k, v)\n",
    "\n",
    "threshold = 0.01\n",
    "for t in ct_cols:\n",
    "    cnt = ms_format[t][ms_format[t] != \"NA\"].value_counts()\n",
    "    cnt_ratio = cnt / cnt.sum()\n",
    "    rm_list = list(cnt_ratio[cnt_ratio < threshold].index)\n",
    "    for r in rm_list:\n",
    "        ms_format[t] = ms_format[t].replace(r, \"NA\")\n",
    "\n",
    "# Save for visualization\n",
    "tax_cols = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "save_cols = ['species_tax_id'] + tax_cols + nt_cols + ct_cols\n",
    "\n",
    "# Categorical trait: binarize and convert 0 or 1\n",
    "binarized_ct_dict = {}\n",
    "rm_set = set([\"NA\", \"Unknown\"])\n",
    "def binarize(x, u):\n",
    "    if x == \"NA\":\n",
    "        res = \"NA\"\n",
    "    elif x == u:\n",
    "        res = 1\n",
    "    else:\n",
    "        res = 0\n",
    "    return res\n",
    "\n",
    "ct_cols = ['gram_stain', 'respiration', 'sporulation', 'motility', 'range_tmp', 'range_salinity', 'cell_shape']\n",
    "# Convert binomial traits to {0,1} values\n",
    "ms_format[\"gram_stain\"] = ms_format[\"gram_stain\"].replace(\"positive\", 1).replace(\"negative\", 0)\n",
    "ms_format[\"sporulation\"] = ms_format[\"sporulation\"].replace(\"yes\", 1).replace(\"no\", 0)\n",
    "ms_format[\"motility\"] = ms_format[\"motility\"].replace(\"yes\", 1).replace(\"no\", 0)\n",
    "ms_format[\"range_salinity\"] = ms_format[\"range_salinity\"].replace(\"halophilic\", 1).replace(\"non-halophilic\", 0)\n",
    "\n",
    "# Convert multinomial traits to three or more traits\n",
    "for t in [\"respiration\", \"range_tmp\", \"cell_shape\"]:\n",
    "    uniq = list(ms_format[t].unique())\n",
    "    uniq = [u for u in uniq if u not in rm_set]\n",
    "    for u in uniq:\n",
    "        binary = ms_format[t].apply(lambda x: binarize(x, u))\n",
    "        binarized_ct_dict[u + \"_\" + t] = list(binary)\n",
    "\n",
    "ms_format_binarized_ct = pd.DataFrame(binarized_ct_dict)\n",
    "\n",
    "# Concat\n",
    "ms_format_binarized = pd.concat([ms_format[[\"species_tax_id\"]+nt_cols+[\"gram_stain\", \"sporulation\", \"motility\", \"range_salinity\"]], ms_format_binarized_ct], axis=1)\n",
    "# Save\n",
    "trait_b2f_path = os.path.join(out_dir, 'ref_bac2feature', 'trait_bac2feature.tsv')\n",
    "taxonomy_path = os.path.join(intermediate_dir, 'silva_taxonomy.tsv')\n",
    "ms_format_binarized.to_csv(trait_b2f_path, sep='\\t', index=False)\n",
    "ms_format[['species_tax_id']+tax_cols].to_csv(taxonomy_path, sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data description of prokaryotic traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "trait = pd.read_csv(trait_b2f_path, sep='\\t')\n",
    "\n",
    "titles = {'cell_diameter': 'Cell diameter', 'cell_length': 'Cell length', 'doubling_h': 'Doubling time', 'growth_tmp': 'Growth temp.', 'optimum_tmp': 'Optimum temp.', 'optimum_ph': 'Optimum pH', 'genome_size': 'Genome size', 'gc_content': 'GC content', 'coding_genes': 'Coding genes', 'rRNA16S_genes': 'rRNA16S genes', 'tRNA_genes': 'tRNA genes', 'gram_stain': 'Gram stain', 'sporulation': 'Sporulation', 'motility': 'Motility', 'range_salinity': 'Salinity range', 'respiration': 'Respiration', 'range_tmp': 'Temperature range', 'cell_shape': 'Cell shape'}\n",
    "\n",
    "units = {'cell_diameter': 'µm (log10)', 'cell_length': 'µm (log10)', 'doubling_h': 'Hours (log10)', 'growth_tmp': 'Degrees C', 'optimum_tmp': 'Degrees C', 'optimum_ph': 'pH', 'genome_size': 'Base pair', 'gc_content': 'Percentage', 'coding_genes': 'Number', 'rRNA16S_genes': 'Number', 'tRNA_genes': 'Number'}\n",
    "\n",
    "# Trait list\n",
    "nt = ['cell_diameter', 'cell_length', 'doubling_h', 'growth_tmp', 'optimum_tmp', 'optimum_ph', 'genome_size', 'gc_content', 'coding_genes', 'rRNA16S_genes', 'tRNA_genes']\n",
    "binary_ct = ['gram_stain', 'sporulation', 'motility', 'range_salinity']\n",
    "multi_ct = ['respiration', 'range_tmp', 'cell_shape']\n",
    "\n",
    "# Plot\n",
    "ntraits = len(nt)+len(binary_ct)+len(multi_ct)\n",
    "ncols = 4\n",
    "nrows = math.ceil(float(ntraits) / ncols)\n",
    "fig, axes = plt.subplots(nrows, ncols, figsize=(2.5*ncols, 2*nrows))\n",
    "\n",
    "# Continuous traits\n",
    "for tr, ax in zip(nt, axes.flatten()):\n",
    "    n, _, _ = ax.hist(trait[tr], bins=15, color='grey', edgecolor='black')\n",
    "    ax.text(1, 1, f'n = {np.sum(~trait[tr].isnull())}',\n",
    "            horizontalalignment='right', verticalalignment='top', transform=ax.transAxes)\n",
    "    ax.set_title(titles[tr])\n",
    "    ax.set_xlabel(units[tr])\n",
    "    ax.set_ylim(0, n.max() * 1.2)\n",
    "    # ax.tick_params(axis=\"x\", labelrotation=90)\n",
    "    ax.spines['top'].set(alpha=0.1, color='grey')\n",
    "    ax.spines['right'].set(alpha=0.1, color='grey')\n",
    "\n",
    "# Binary traits\n",
    "for tr, ax in zip(binary_ct, axes.flatten()[len(nt):]):\n",
    "    trait_values = trait[tr].dropna()\n",
    "    if tr == 'gram_stain':\n",
    "        trait_values = trait_values.replace(0, 'no').replace(1, 'yes')\n",
    "    if tr == 'sporulation':\n",
    "        trait_values = trait_values.replace(0, 'negative').replace(1, 'positive')\n",
    "    if tr == 'motility':\n",
    "        trait_values = trait_values.replace(0, 'no').replace(1, 'yes')\n",
    "    if tr == 'range_salinity':\n",
    "        trait_values = trait_values.replace(0, 'non-halophilic').replace(1, 'halophilic')\n",
    "    cnt = Counter(trait_values)\n",
    "    x = np.array(list(cnt.keys()))\n",
    "    height = np.array(list(cnt.values()))\n",
    "    sorted_index = np.argsort(height)[::-1]\n",
    "    ax.bar(x=x[sorted_index], height=np.sort(height)[::-1], color='grey', edgecolor='black')\n",
    "    ax.text(1, 1, f'n = {np.sum(~trait[tr].isnull())}',\n",
    "            horizontalalignment='right', verticalalignment='top', transform=ax.transAxes)\n",
    "    ax.set_title(titles[tr])\n",
    "    ax.set_ylim(0, height.max() * 1.2)\n",
    "    ax.tick_params(axis=\"x\", labelrotation=90)\n",
    "    # ax.set_xlabel(nt_units[tr])\n",
    "    ax.spines['top'].set(alpha=0.1, color='grey')\n",
    "    ax.spines['right'].set(alpha=0.1, color='grey')\n",
    "\n",
    "# Multinomial traits\n",
    "cnt_multi_ct = dict()\n",
    "for mc in multi_ct:\n",
    "    cnt = dict()\n",
    "    for tr in trait.columns[16:]:\n",
    "        if mc in tr:\n",
    "            trait_values = trait[tr].dropna()\n",
    "            trait_values = trait_values.replace(0, 'no').replace(1, tr)\n",
    "            trait_category = tr.replace('_'+mc, '')\n",
    "            cnt[trait_category] = Counter(trait_values)[tr]\n",
    "    cnt_multi_ct[mc] = cnt\n",
    "\n",
    "for tr, ax in zip(multi_ct, axes.flatten()[len(nt)+len(binary_ct):]):\n",
    "    x = np.array(list(cnt_multi_ct[tr].keys()))\n",
    "    height = np.array(list(cnt_multi_ct[tr].values()))\n",
    "    sorted_index = np.argsort(height)[::-1]\n",
    "    ax.bar(x=x[sorted_index], height=np.sort(height)[::-1], color='grey', edgecolor='black')\n",
    "    ax.text(1, 1, f'n = {height.sum()}',\n",
    "            horizontalalignment='right', verticalalignment='top', transform=ax.transAxes)\n",
    "    ax.set_ylim(0, height.max() * 1.2)\n",
    "    ax.set_title(titles[tr])\n",
    "    ax.tick_params(axis=\"x\", labelrotation=90)\n",
    "    ax.spines['top'].set(alpha=0.1, color='grey')\n",
    "    ax.spines['right'].set(alpha=0.1, color='grey')\n",
    "\n",
    "for i in range(ntraits, nrows*ncols):\n",
    "    axes.flatten()[i].set_visible(False)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save\n",
    "plt.savefig(\"../../results/02_data_preprocessing/figS1.pdf\", \n",
    "            format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare reference datasets for Bac2Feature"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Homology-based prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extracting 16S rRNA seqs from SILVA SSU Ref NR99"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the remaining node ID\n",
    "acc_to_taxid = {madin_silva.loc[i, 'accession']: madin_silva.loc[i, 'species_tax_id'] for i in range(madin_silva.shape[0])}\n",
    "\n",
    "# Extract full-length 16S rRNA sequences\n",
    "silva_seq_path = os.path.join(silva_dir, 'SILVA_138.1_SSURef_NR99_tax_silva.fasta')\n",
    "silva_seq_id_path = os.path.join(intermediate_dir, 'SILVA_138.1_SSURef_NR99_tax_silva_taxid.fasta')\n",
    "\n",
    "records = list(SeqIO.parse(silva_seq_path, format='fasta'))\n",
    "results_records = [r for r in records if r.id in acc_to_taxid]\n",
    "\n",
    "for r in results_records:\n",
    "    r.seq = r.seq.replace('.', '-')\n",
    "    # Convert from RNA to DNA\n",
    "    r.seq = r.seq.replace('U', 'T')\n",
    "    # Convert from accession to taxid\n",
    "    r.id = acc_to_taxid[r.id]\n",
    "    r.description = \"\"\n",
    "\n",
    "SeqIO.write(results_records, silva_seq_id_path, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Constructing BLAST Database\n",
    "\n",
    "```sh\n",
    "makeblastdb -in ../../data/intermediate_dir/SILVA_138.1_SSURef_NR99_tax_silva_taxid.fasta -out ../../data/ref_bac2feature/homology/blastdb -dbtype nucl\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Taxonomy-based prediction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Calculating empirical distributions of reference trait data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make reference dataset for taxonomy-based prediction\n",
    "\n",
    "# Trait table\n",
    "trait = pd.read_csv(trait_b2f_path, sep=\"\\t\", dtype=str)\n",
    "# SILVA taxonomy\n",
    "silva_taxonomy = pd.read_csv(taxonomy_path, sep=\"\\t\", dtype=str)\n",
    "silva_taxonomy.fillna(\"\", inplace=True)\n",
    "\n",
    "def concat_taxonomy(silva_metadata):\n",
    "    taxonomy = (\"k__\" + silva_metadata[\"superkingdom\"] +\n",
    "                \"; p__\" + silva_metadata[\"phylum\"] +\n",
    "                \"; c__\" + silva_metadata[\"class\"] +\n",
    "                \"; o__\" + silva_metadata[\"order\"] +\n",
    "                \"; f__\" + silva_metadata[\"family\"] +\n",
    "                \"; g__\" + silva_metadata[\"genus\"] +\n",
    "                \"; s__\" + silva_metadata[\"species\"])\n",
    "    return taxonomy\n",
    "\n",
    "silva_taxonomy[\"taxonomy\"] = silva_taxonomy.apply(lambda df: concat_taxonomy(df), axis=1)\n",
    "# Save\n",
    "silva_taxonomy[[\"species_tax_id\", \"taxonomy\"]].to_csv(os.path.join(out_dir, 'intermediate_dir', 'qiime_taxonomy.tsv'), sep='\\t', index=False)\n",
    "\n",
    "# Merge\n",
    "tax_cols = ['taxonomy', 'superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']\n",
    "ref_trait = pd.merge(trait, silva_taxonomy[[\"species_tax_id\"]+tax_cols], how=\"inner\", on=\"species_tax_id\")\n",
    "\n",
    "nt = ['cell_diameter', 'cell_length', 'doubling_h', 'growth_tmp', 'optimum_tmp', 'optimum_ph', 'genome_size', 'gc_content', 'coding_genes', 'rRNA16S_genes', 'tRNA_genes']\n",
    "\n",
    "ct = ['gram_stain',\n",
    "      'sporulation', 'motility', 'range_salinity', 'facultative_respiration',\n",
    "      'anaerobic_respiration', 'aerobic_respiration', 'mesophilic_range_tmp',\n",
    "      'thermophilic_range_tmp', 'psychrophilic_range_tmp',\n",
    "      'bacillus_cell_shape', 'coccus_cell_shape', 'filament_cell_shape',\n",
    "      'coccobacillus_cell_shape', 'vibrio_cell_shape', 'spiral_cell_shape']\n",
    "\n",
    "clades = [\"superkingdom\", \"phylum\", \"class\",\n",
    "          \"order\", \"family\", \"genus\", \"species\"]\n",
    "prefix = [\"k__\", \"p__\", \"c__\", \"o__\", \"f__\", \"g__\", \"s__\"]\n",
    "\n",
    "# Calculate empirical distributions of reference traits\n",
    "emp_trait = {}\n",
    "for t in nt+ct:\n",
    "    emp = {}\n",
    "    ref_trait[t] = ref_trait[t].astype(float)\n",
    "    for c in clades:\n",
    "        cols = [c, t]\n",
    "        groupby = ref_trait[cols].groupby(c).mean()\n",
    "        groupby = groupby.reset_index()\n",
    "        groupby.fillna({t: \"NA\"}, inplace=True)\n",
    "        emp[c] = {groupby.loc[i, c]: groupby.loc[i, t]\n",
    "                    for i in range(groupby.shape[0])}\n",
    "    emp_trait[t] = emp\n",
    "# Save\n",
    "emp_file = open(f'../../data/ref_bac2feature/taxonomy/empirical_dictribution.json', mode='w')\n",
    "json.dump(emp_trait, emp_file, ensure_ascii=False)\n",
    "emp_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Training naive-bayes classifier\n",
    "```sh\n",
    "# Activate Conda environment\n",
    "conda activate qiime2-2023.5\n",
    "# Import\n",
    "qiime tools import --type 'FeatureData[Taxonomy]' --input-format HeaderlessTSVTaxonomyFormat --input-path ../../data/intermediate_dir/qiime_taxonomy.tsv --output-path ../../data/intermediate_dir/qiime_taxonomy.qza\n",
    "qiime tools import --type 'FeatureData[Sequence]' --input-path ../../data/intermediate_dir/SILVA_138.1_SSURef_NR99_tax_silva_taxid.fasta --output-path ../../data/intermediate_dir/ref_seq_full.qza\n",
    "# Train naive-bayes classifier\n",
    "qiime feature-classifier fit-classifier-naive-bayes --i-reference-reads ../../data/intermediate_dir/ref_seq_full.qza --i-reference-taxonomy ../../data/intermediate_dir/qiime_taxonomy.tsv --o-classifier ../../data/ref_bac2feature/taxonomy/nb_classifier.qza\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phylogeny-based prediction\n",
    "\n",
    "```sh\n",
    "conda activate bac2feature_experiment\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pruning the phylogenetic tree of SILVA SSU Ref NR99"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Path\n",
    "silva_large_tree_w_single_quote_path = os.path.join(silva_dir, 'SILVA_138.1_SSURef_NR99.accessions.ntree')\n",
    "silva_large_tree_path = os.path.join(intermediate_dir, 'silva_large_tree.tre')\n",
    "index_list_path = os.path.join(intermediate_dir, 'remained_nodeid_index.txt')\n",
    "silva_pruned_tree_path = os.path.join(intermediate_dir, 'silva_pruned_tree.tre')\n",
    "silva_pruned_taxid_tree_path = os.path.join(intermediate_dir, 'silva_pruned_taxid_tree.tre')\n",
    "\n",
    "# Remove single quotation to avoid an read error\n",
    "tree = Phylo.read(silva_large_tree_w_single_quote_path, format='newick')\n",
    "leaves = tree.get_terminals()\n",
    "sys.setrecursionlimit(10000)\n",
    "Phylo.write(tree, silva_large_tree_path, 'newick')\n",
    "\n",
    "# Save the remaining node ID\n",
    "acc_to_taxid = {madin_silva.loc[i, 'accession']: madin_silva.loc[i, 'species_tax_id'] for i in range(madin_silva.shape[0])}\n",
    "\n",
    "remained_nodeid = [l for l in leaves if l.name in acc_to_taxid]\n",
    "with open(index_list_path, 'w') as f:\n",
    "    f.writelines([str(id) + '\\n' for id in remained_nodeid])\n",
    "\n",
    "# Pruning the phylogenetic tree\n",
    "command = ['Rscript',\n",
    "           'prune_tree.R',\n",
    "           '-i',\n",
    "           silva_large_tree_path,\n",
    "           '-o',\n",
    "           silva_pruned_tree_path,\n",
    "           '-k',\n",
    "           index_list_path]\n",
    "process = subprocess.run(command)\n",
    "\n",
    "# Preprocess the pruned tree\n",
    "pruned_tree = Phylo.read(silva_pruned_tree_path, format='newick')\n",
    "# To input into RAxML, the node labels must be removed.\n",
    "for t in pruned_tree.get_nonterminals():\n",
    "    t.name = ''\n",
    "# Convert accession to NCBI taxid\n",
    "for t in pruned_tree.get_terminals():\n",
    "    t.name = acc_to_taxid[t.name]\n",
    "\n",
    "# Save\n",
    "Phylo.write(pruned_tree, silva_pruned_taxid_tree_path, 'newick')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extracting 16S rRNA seqs from SILVA MSA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract MSA of 16S rRNA sequences\n",
    "silva_seq_path = os.path.join(silva_dir, 'SILVA_138.1_SSURef_NR99_tax_silva_full_align_trunc.fasta')\n",
    "silva_seq_id_path = os.path.join(intermediate_dir, 'SILVA_138.1_SSURef_NR99_tax_silva_full_align_trunc_taxid.fasta')\n",
    "\n",
    "records = list(SeqIO.parse(silva_seq_path, format='fasta'))\n",
    "results_records = [r for r in records if r.id in acc_to_taxid]\n",
    "\n",
    "for r in results_records:\n",
    "    r.seq = r.seq.replace('.', '-')\n",
    "    # Convert from RNA to DNA\n",
    "    r.seq = r.seq.replace('U', 'T')\n",
    "    # Convert from accession to taxid\n",
    "    r.id = acc_to_taxid[r.id]\n",
    "    r.description = \"\"\n",
    "\n",
    "SeqIO.write(results_records, silva_seq_id_path, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Trim MSA of 16S rRNA gene sequences with trimal\n",
    "```sh\n",
    "trimal -gt 0.10 -in ../../data/intermediate_dir/SILVA_138.1_SSURef_NR99_tax_silva_full_align_trunc_taxid.fasta -out ../../data/intermediate_dir/silva_msa_trim.fasta\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Construct of the reference dataset for PICRUSt2's pipeline\n",
    "```sh\n",
    "# MSA\n",
    "cp ../../data/intermediate_dir/silva_msa_trim.fasta ../../data/ref_bac2feature/phylogeny/phylogeny.fasta\n",
    "# HMM profile\n",
    "hmmbuild ../../data/ref_bac2feature/phylogeny/phylogeny.hmm ../../data/ref_bac2feature/phylogeny/phylogeny.fasta\n",
    "# Parameter estimation by Raxml-ng\n",
    "raxml-ng --evaluate --msa ../../data/ref_bac2feature/phylogeny/phylogeny.fasta --tree ../../data/intermediate_dir/silva_pruned_taxid_tree.tre --prefix ../../data/intermediate_dir/phylogeny --model GTR+G+F --blmin 0.0001 -blmax 1.5\n",
    "# Rerooting tree optimized by Raxml-ng (root ad midpoint)\n",
    "Rscript rerooting.R\n",
    "# Copy tree and model\n",
    "cp ../../data/intermediate_dir/phylogeny.raxml.bestTree.rerooted ../../data/ref_bac2feature/phylogeny/phylogeny.tre\n",
    "cp ../../data/intermediate_dir/phylogeny.raxml.bestModel ../../data/ref_bac2feature/phylogeny/phylogeny.model\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bac2feature",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
