{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare trait and 16S rRNA gene sequence data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment settings\n",
    "```sh\n",
    "# Working Directory\n",
    "cd Bac2fFeature/scripts/01_data_preparation\n",
    "# Conda environment\n",
    "conda create --name bac2feature_experiment --file ../../environment/env_bac2feature_experiment.yml\n",
    "conda create --name qiime2-2023.5 --file ../../environment/env_qiime2-2023.5-py38-linux-conda.yml\n",
    "# Activate Conda environment\n",
    "conda activate bac2feature_experiment\n",
    "# Output Directory\n",
    "directories=(\n",
    "    \"../../data/\"\n",
    "    \"../../data/silva\"\n",
    "    \"../../data/trait\"\n",
    "    \"../../data/intermediate_dir\"\n",
    ")\n",
    "for dir in \"${directories[@]}\"; do\n",
    "  if [ ! -d \"$dir\" ]; then\n",
    "    mkdir -p \"$dir\"\n",
    "  fi\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trait database (Madin et al., 2020)\n",
    "Downloaded at 2024.06.06\n",
    "```sh\n",
    "wget -nc -P ../../data/trait https://raw.githubusercontent.com/bacteria-archaea-traits/bacteria-archaea-traits/master/output/condensed_species_NCBI.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SILVA 16S rRNA seqs (version 138.1)\n",
    "Downloaded at 2024.06.06\n",
    "```sh\n",
    "silva_dir=\"../../data/silva\"\n",
    "# 16S rRNA sequences, multiple sequence alignment, phylogenetic tree, and metadata from SILVA release. 138\n",
    "## Full 16S rRNA sequences\n",
    "wget -nc -P ${silva_dir} https://www.arb-silva.de/fileadmin/silva_databases/release_138.1/Exports/SILVA_138.1_SSURef_NR99_tax_silva.fasta.gz\n",
    "## Multiple sequence alignment\n",
    "wget -nc -P ${silva_dir} https://www.arb-silva.de/fileadmin/silva_databases/release_138.1/Exports/SILVA_138.1_SSURef_NR99_tax_silva_full_align_trunc.fasta.gz\n",
    "## Phylogenetic tree derived from 16S rRNA seqs\n",
    "wget -nc -P ${silva_dir} https://www.arb-silva.de/fileadmin/silva_databases/release_138.1/user/guide-trees/SILVA_138.1_SSURef_NR99.accessions.ntree.gz\n",
    "## Metadata\n",
    "wget -nc -P ${silva_dir} https://www.arb-silva.de/fileadmin/silva_databases/release_138.1/Exports/full_metadata/SILVA_138.1_SSURef_Nr99.full_metadata.gz\n",
    "# Unpack\n",
    "find ${silva_dir} -type f -name \"*.gz\" -exec unpigz -k {} \\;\n",
    "```"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bac2feature",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
