{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Visualization: compare_hsp_method.ipynb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment settings\n",
    "```sh\n",
    "# Working Directory\n",
    "cd Bac2fFeature/scripts/09_cross_validation_suppl/093_phylogeny\n",
    "# Output Directory\n",
    "directories=(\n",
    "    \"../../../data/cross_validation_suppl\"\n",
    ")\n",
    "for dir in \"${directories[@]}\"; do\n",
    "  if [ ! -d \"$dir\" ]; then\n",
    "    mkdir -p \"$dir\"\n",
    "  fi\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "matplotlib.rcParams['font.family']       = 'Arial'\n",
    "matplotlib.rcParams['font.sans-serif']   = [\"Arial\",\"DejaVu Sans\",\"Lucida Grande\",\"Verdana\"]\n",
    "matplotlib.rcParams['figure.figsize']    = [4,3]\n",
    "matplotlib.rcParams['font.size']         = 10\n",
    "matplotlib.rcParams[\"axes.labelcolor\"]   = \"#000000\"\n",
    "matplotlib.rcParams[\"axes.linewidth\"]    = 1.0\n",
    "matplotlib.rcParams[\"xtick.major.width\"] = 1.0\n",
    "matplotlib.rcParams[\"ytick.major.width\"] = 1.0\n",
    "cmap1 = plt.cm.tab20\n",
    "cmap2 = plt.cm.Set3\n",
    "\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_xy_vals(cmp, t):\n",
    "    known_flag = (~cmp[t+'_t'].isnull()) & (~cmp[t+'_e'].isnull())\n",
    "    estimated_vals, true_vals = cmp[known_flag][t+'_e'], cmp[known_flag][t+'_t']\n",
    "    return estimated_vals, true_vals\n",
    "\n",
    "def calc_cor(df, t):\n",
    "    estimated_vals, true_vals = get_xy_vals(df, t)\n",
    "    return estimated_vals.corr(true_vals)\n",
    "\n",
    "def calc_macro_f1(df, t):\n",
    "    estimated_vals, true_vals = get_xy_vals(df, t)\n",
    "    # calc confusion matrix\n",
    "    category = list(true_vals.unique())\n",
    "    cnt = pd.DataFrame(data={i: [0] * len(category) for i in category},\n",
    "                    index=category,\n",
    "                    columns=category\n",
    "                    )\n",
    "    for e, t in zip(estimated_vals, true_vals):\n",
    "        cnt[e][t] += 1\n",
    "    cnt = cnt.apply(lambda x: x/x.sum(), axis=1)\n",
    "    # calc macro-F1\n",
    "    macro_f1 = np.diag(cnt).mean()\n",
    "    return macro_f1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "titles = {'cell_diameter': 'Cell diameter', 'cell_length': 'Cell length', 'doubling_h': 'Doubling time', 'growth_tmp': 'Growth temp.', 'optimum_tmp': 'Optimum temp.', 'optimum_ph': 'Optimum pH', 'genome_size': 'Genome size', 'gc_content': 'GC content', 'coding_genes': 'Coding genes', 'rRNA16S_genes': 'rRNA16S genes', 'tRNA_genes': 'tRNA genes', 'gram_stain': 'Gram stain', 'sporulation': 'Sporulation', 'motility': 'Motility', 'range_salinity': 'Halophile', 'facultative_respiration': 'Facultative', 'anaerobic_respiration': 'Anaerobe', 'aerobic_respiration':'Aerobe' ,'mesophilic_range_tmp': 'Mesophile', 'thermophilic_range_tmp':'Thermophile', 'psychrophilic_range_tmp': 'Psychrophile', 'bacillus_cell_shape': 'Bacillus', 'coccus_cell_shape': 'Coccus', 'filament_cell_shape': 'Filament', 'coccobacillus_cell_shape': 'Coccobacillus', 'vibrio_cell_shape': 'Vibrio', 'spiral_cell_shape': 'Spiral'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Continuous traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dic_method_result = {\"WSCP\": \"hsp_results/hsp_result_wscp.tsv\",\n",
    "                     \"PIC\":  \"hsp_results/hsp_result_pic.tsv\",\n",
    "                     \"SA\":   \"hsp_results/hsp_result_sa.tsv\"}\n",
    "\n",
    "nt = ['cell_diameter', 'cell_length', 'doubling_h', 'genome_size', 'gc_content', 'coding_genes', 'optimum_tmp', 'optimum_ph', 'growth_tmp', 'rRNA16S_genes', 'tRNA_genes']\n",
    "\n",
    "fig, axes = plt.subplots(3, 4, figsize=(8, 6))\n",
    "\n",
    "\n",
    "trait_path = \"../../../data/ref_bac2feature/trait_bac2feature.tsv\"\n",
    "true_vals = pd.read_csv(trait_path, sep=\"\\t\")\n",
    "\n",
    "for k, hsp_result_path in dic_method_result.items():\n",
    "    for i, t in enumerate(nt):\n",
    "\n",
    "        estimated_vals = pd.read_csv(hsp_result_path, sep=\"\\t\")\n",
    "\n",
    "        cmp = pd.merge(estimated_vals, true_vals, left_on=\"species_tax_id\", right_on=\"species_tax_id\", how=\"inner\", suffixes=[\"_e\", \"_t\"])\n",
    "\n",
    "        score = cmp.groupby(\"threshold\").apply(lambda df: calc_cor(df, t))\n",
    "        threshold = cmp[\"threshold\"].unique()\n",
    "\n",
    "        ax = axes.flatten()[i]\n",
    "        ax.plot(threshold, score, marker='+')\n",
    "\n",
    "        ax.set_ylim(0, 1)\n",
    "        ax.set_xlabel(\"Phylogenetic distance\")\n",
    "        ax.set_ylabel(\"Pearson's r\")\n",
    "        ax.set_title(titles[t])\n",
    "\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "\n",
    "# axes.flatten()[0].legend([\"WSCP\", \"PIC\", \"SA\"])\n",
    "axes.flatten()[-1].set_visible(False)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(\"../../../results/09_cross_validation_suppl/figS4a.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Categorical traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dic_method_result = {\n",
    "                     \"EMP\": \"hsp_results//hsp_result_emp.tsv\",\n",
    "                     \"Mk-SYM\":   \"hsp_results/hsp_result_mk_SYM.tsv\",\n",
    "                     \"Mk-ARD\":  \"hsp_results/hsp_result_mk_ARD.tsv\",\n",
    "                     \"MP-no-edge\":  \"hsp_results/hsp_result_mp_no_edge.tsv\",\n",
    "                     \"MP-edge\":  \"hsp_results/hsp_result_mp_inversed_edge.tsv\"\n",
    "                     }\n",
    "\n",
    "ct = [\"gram_stain\", \"sporulation\", \"motility\", \"range_salinity\", \"facultative_respiration\", \"anaerobic_respiration\", \"aerobic_respiration\", \"mesophilic_range_tmp\", \"thermophilic_range_tmp\", \"psychrophilic_range_tmp\", \"bacillus_cell_shape\", \"coccus_cell_shape\", \"filament_cell_shape\", \"coccobacillus_cell_shape\", \"vibrio_cell_shape\", \"spiral_cell_shape\"]\n",
    "\n",
    "ncols = 4\n",
    "nrows = math.ceil(float(len(ct)) / ncols)\n",
    "fig, axes = plt.subplots(nrows, ncols, figsize=(2*ncols, 2*nrows))\n",
    "\n",
    "trait_path = \"../../../data/ref_bac2feature/trait_bac2feature.tsv\"\n",
    "true_vals = pd.read_csv(trait_path, sep=\"\\t\")\n",
    "\n",
    "for k, hsp_result_path in dic_method_result.items():\n",
    "    for i, t in enumerate(ct):\n",
    "\n",
    "        estimated_vals = pd.read_csv(hsp_result_path, sep=\"\\t\")\n",
    "\n",
    "        cmp = pd.merge(estimated_vals, true_vals, left_on=\"species_tax_id\", right_on=\"species_tax_id\", how=\"inner\", suffixes=[\"_e\", \"_t\"])\n",
    "\n",
    "        score = cmp.groupby(\"threshold\").apply(lambda df: calc_cor(df, t))\n",
    "        threshold = cmp[\"threshold\"].unique()\n",
    "\n",
    "        ax = axes.flatten()[i]\n",
    "        ax.plot(threshold, score, marker='+')\n",
    "\n",
    "        ax.set_ylim(-0.3, 1)\n",
    "        ax.set_xlabel(\"Phylogenetic distance\")\n",
    "        ax.set_ylabel(\"MCC\")\n",
    "        ax.set_title(titles[t])\n",
    "\n",
    "        ax.spines['top'].set_visible(False)\n",
    "        ax.spines['right'].set_visible(False)\n",
    "\n",
    "# axes.flatten()[0].legend([\"EMP\", \"Mk-ER\", \"Mk-SYM\", \"MP-no-edge\", \"MP-edge\"])\n",
    "axes.flatten()[-1].set_visible(False)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(\"../../../results/09_cross_validation_suppl/figS4b.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bac2feature",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
