{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculating phylogenetic thresholds and applying them to EMP dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment settings\n",
    "```sh\n",
    "# Working Directory\n",
    "cd Bac2fFeature/scripts/06_trait_autocorrelations\n",
    "# Output Directory\n",
    "directories=(\n",
    "    \"../../data/trait_autocorrelations\"\n",
    ")\n",
    "for dir in \"${directories[@]}\"; do\n",
    "  if [ ! -d \"$dir\" ]; then\n",
    "    mkdir -p \"$dir\"\n",
    "  fi\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating phylogenetic distance where traits are correlated\n",
    "```sh\n",
    "# Use Conda environment for R\n",
    "conda activate r_bac2feature\n",
    "# Calculation of trait autocorrelation along with phylogenetic signals\n",
    "Rscript calc_autocorrelations.R\n",
    "# Setting thresholds\n",
    "Rscript setting_phylogenetic_signals.R\n",
    "# Use default Conda environment in the following analysis\n",
    "conda activate bac2feature_experiment\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate phylogenetic signals in Earth Microbiome Project1 dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from matplotlib.ticker import FixedLocator, FormatStrFormatter\n",
    "from matplotlib.gridspec import GridSpec, GridSpecFromSubplotSpec\n",
    "\n",
    "matplotlib.rcParams['font.family']       = 'Arial'\n",
    "matplotlib.rcParams['font.sans-serif']   = [\"Arial\",\"DejaVu Sans\",\"Lucida Grande\",\"Verdana\"]\n",
    "matplotlib.rcParams[\"figure.figsize\"]    = [4,3]\n",
    "matplotlib.rcParams[\"font.size\"]         = 10\n",
    "matplotlib.rcParams[\"axes.labelcolor\"]   = \"#000000\"\n",
    "matplotlib.rcParams[\"axes.linewidth\"]    = 1.0\n",
    "matplotlib.rcParams[\"xtick.major.width\"] = 1.0\n",
    "matplotlib.rcParams[\"ytick.major.width\"] = 1.0\n",
    "cmap1 = plt.cm.tab20\n",
    "cmap2 = plt.cm.Set3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Preprocessing EMP data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert tsv to fasta\n",
    "otu_table_filepath = \"../../data/trait_autocorrelations/emp_deblur_150bp.subset_2k.rare_5000.tsv\"\n",
    "otu_fasta_filepath = \"../../data/trait_autocorrelations/otu_emp_deblur_150bp.subset_2k.rare_5000.fasta\"\n",
    "\n",
    "otu_table = pd.read_csv(otu_table_filepath, sep=\"\\t\", header=1)\n",
    "otu_table[\"#OTU No\"] = [\"OTU\" + str(i) for i in np.arange(0, otu_table.shape[0])]\n",
    "\n",
    "def make_seqrecord(df):\n",
    "    return SeqRecord(\n",
    "        Seq(df[\"#OTU ID\"]),\n",
    "        id=df[\"#OTU No\"],\n",
    "        name=\"\",\n",
    "        description=\"\",\n",
    "    )\n",
    "records = otu_table.apply(lambda df: make_seqrecord(df), axis=1)\n",
    "\n",
    "with open(otu_fasta_filepath, \"w\") as outfh:\n",
    "    SeqIO.write(records, outfh, \"fasta\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculating phylogenetic distance\n",
    "```sh\n",
    "bac2feature -s ../../data/trait_autocorrelations/otu_emp_deblur_150bp.subset_2k.rare_5000.fasta -o ../../data/trait_autocorrelations/bac2feature_result.tsv -m phylogeny --ref_dir ../../data/ref_bac2feature/phylogeny --ref_trait ../../data/ref_bac2feature/trait_bac2feature.tsv --intermediate_dir ../../data/trait_autocorrelations/intermediate_dir --calculate_NSTI\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Calculating relative abundance of ASVs within thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Phylogenetic distance\n",
    "b2f_result = pd.read_csv(\"../../data/trait_autocorrelations/bac2feature_result.tsv\", sep=\"\\t\")\n",
    "\n",
    "# EMP dataset\n",
    "otu_table_filepath = \"../../data/trait_autocorrelations/emp_deblur_150bp.subset_2k.rare_5000.tsv\"\n",
    "otu_table = pd.read_csv(otu_table_filepath, sep=\"\\t\", header=1)\n",
    "\n",
    "all_sample_list = list(otu_table.columns[1:])\n",
    "otu_table[\"sequence\"] = [\"OTU\" + str(i) for i in np.arange(0, otu_table.shape[0])]\n",
    "\n",
    "metadata_filepath = \"../../data/trait_autocorrelations/emp_qiime_mapping_subset_2k.tsv\"\n",
    "metadata = pd.read_csv(metadata_filepath, sep=\"\\t\", header=0, index_col=0)\n",
    "metadata = metadata.loc[all_sample_list,:]\n",
    "metadata = metadata.reset_index()\n",
    "\n",
    "sample_with_metadata = list(metadata[\"#SampleID\"])\n",
    "\n",
    "total_abandance = otu_table.loc[:, sample_with_metadata].sum(axis=0)\n",
    "relative_abundance = otu_table.loc[:, sample_with_metadata] / total_abandance\n",
    "otu_relative_abundance = pd.concat([otu_table.loc[:, \"sequence\"], relative_abundance], axis=1)\n",
    "\n",
    "rel_abd_melt = otu_relative_abundance.melt(id_vars=\"sequence\", var_name=\"#SampleID\", value_name=\"abundance\")\n",
    "rel_abd_meta = pd.merge(rel_abd_melt, metadata[[\"#SampleID\", \"empo_2\", \"empo_3\"]], how=\"left\", on=\"#SampleID\")\n",
    "\n",
    "# Threshold\n",
    "threshold_phylodistance = pd.read_csv(\"../../data/trait_autocorrelations/threshold_phylodistance.tsv\", sep= \"\\t\", index_col=0)\n",
    "threshold_phylodistance.rename(columns={\"cor_0.5\": \"threshold\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def abd_above_threshold(df, t, threshold):\n",
    "    return df[\"abundance\"] if df[t+\"_nsti\"] < threshold else 0\n",
    "\n",
    "# Continuous traits \n",
    "# Remove traits with thresholds values 0\n",
    "trait_list = [\"doubling_h\", \"genome_size\", \"gc_content\", \"coding_genes\", \"optimum_tmp\", \"growth_tmp\", \"rRNA16S_genes\", \"tRNA_genes\"]\n",
    "\n",
    "env_col = \"empo_2\"\n",
    "above_abd_env_type_2d = []\n",
    "\n",
    "for t in trait_list:\n",
    "    # Calculate weighted NSTI per OTU\n",
    "    abd_nsti = pd.merge(rel_abd_melt, b2f_result[[\"sequence\", t+\"_nsti\"]], how=\"left\", on=\"sequence\")\n",
    "    # extract abundance with nsti is smaller than threshold\n",
    "    threshold = threshold_phylodistance.loc[t, \"threshold\"]\n",
    "    abd_nonzero_nsti = abd_nsti[abd_nsti[\"abundance\"] > 0].copy()\n",
    "    abd_nonzero_nsti[\"above_abd\"] = abd_nonzero_nsti.apply(\n",
    "        lambda df: abd_above_threshold(df, t, threshold), axis=1)\n",
    "\n",
    "    # sum abundance by each sample\n",
    "    above_abd = abd_nonzero_nsti[[\"#SampleID\", \"above_abd\"]].groupby(by=\"#SampleID\").sum()\n",
    "    above_abd = above_abd.reset_index()\n",
    "    above_abd_meta = pd.merge(above_abd, metadata, how=\"left\", on=\"#SampleID\")\n",
    "\n",
    "    # mean by env_type\n",
    "    above_abd_env_type = above_abd_meta[[\"above_abd\", env_col]].groupby(by=env_col).mean()\n",
    "    above_abd_env_type = above_abd_env_type.reset_index()\n",
    "\n",
    "    above_abd_env_type_2d.append(list(above_abd_env_type[\"above_abd\"]))\n",
    "\n",
    "df_above_abd_empo2_nt = pd.DataFrame(above_abd_env_type_2d, index=trait_list, columns=list(above_abd_env_type[\"empo_2\"]))\n",
    "\n",
    "# Categorical traits\n",
    "# Remove traits with thresholds values 0\n",
    "trait_list = ['gram_stain', 'sporulation', 'motility', 'anaerobic_respiration', 'mesophilic_range_tmp', 'thermophilic_range_tmp', 'bacillus_cell_shape', 'coccus_cell_shape', 'filament_cell_shape', 'spiral_cell_shape']\n",
    "\n",
    "env_col = \"empo_2\"\n",
    "above_abd_env_type_2d = []\n",
    "\n",
    "for t in trait_list:\n",
    "    # Calculate weighted NSTI per OTU\n",
    "    abd_nsti = pd.merge(rel_abd_melt, b2f_result[[\"sequence\", t+\"_nsti\"]], how=\"left\", on=\"sequence\")\n",
    "    # extract abundance with nsti is smaller than threshold\n",
    "    threshold = threshold_phylodistance.loc[t, \"threshold\"]\n",
    "    abd_nonzero_nsti = abd_nsti[abd_nsti[\"abundance\"] > 0].copy()\n",
    "    abd_nonzero_nsti[\"above_abd\"] = abd_nonzero_nsti.apply(\n",
    "        lambda df: abd_above_threshold(df, t, threshold), axis=1)\n",
    "\n",
    "    # sum abundance by each sample\n",
    "    above_abd = abd_nonzero_nsti[[\"#SampleID\", \"above_abd\"]].groupby(by=\"#SampleID\").sum()\n",
    "    above_abd = above_abd.reset_index()\n",
    "    above_abd_meta = pd.merge(above_abd, metadata, how=\"left\", on=\"#SampleID\")\n",
    "\n",
    "    # mean by env_type\n",
    "    above_abd_env_type = above_abd_meta[[\"above_abd\", env_col]].groupby(by=env_col).mean()\n",
    "    above_abd_env_type = above_abd_env_type.reset_index()\n",
    "\n",
    "    above_abd_env_type_2d.append(list(above_abd_env_type[\"above_abd\"]))\n",
    "\n",
    "df_above_abd_empo2_ct = pd.DataFrame(above_abd_env_type_2d, index=trait_list, columns=list(above_abd_env_type[\"empo_2\"]))\n",
    "\n",
    "# Save\n",
    "df_out = pd.concat([df_above_abd_empo2_nt, df_above_abd_empo2_ct], axis=0)\n",
    "df_out.reset_index(names=['trait'], inplace=True)\n",
    "df_out.to_csv('../../data/trait_autocorrelations/above_abd_empo2.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Visualization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_out = pd.read_csv('../../data/trait_autocorrelations/above_abd_empo2.tsv', sep='\\t', index_col=['trait'])\n",
    "\n",
    "phylod_df = threshold_phylodistance.reset_index(drop=False)\n",
    "empo_df = df_out.reset_index(drop=False)\n",
    "phylod_empo = pd.merge(phylod_df[[\"trait\", \"threshold\"]], empo_df, how='inner', on='trait')\n",
    "\n",
    "nt = ['doubling_h', 'optimum_tmp', 'tRNA_genes', 'rRNA16S_genes', 'coding_genes', 'genome_size', 'growth_tmp', 'gc_content']\n",
    "phylod_empo_nt = phylod_empo.set_index('trait').loc[nt, :]\n",
    "\n",
    "ct = ['mesophilic_range_tmp', 'motility', 'bacillus_cell_shape', 'thermophilic_range_tmp', 'filament_cell_shape', 'spiral_cell_shape', 'coccus_cell_shape', 'anaerobic_respiration', 'sporulation', 'gram_stain']\n",
    "phylod_empo_ct = phylod_empo.set_index('trait').loc[ct, :]\n",
    "\n",
    "titles = {'cell_diameter': 'Cell diameter', 'cell_length': 'Cell length', 'doubling_h': 'Doubling time', 'growth_tmp': 'Growth temp.', 'optimum_tmp': 'Optimum temp.', 'optimum_ph': 'Optimum pH', 'genome_size': 'Genome size', 'gc_content': 'GC content', 'coding_genes': 'Coding genes', 'rRNA16S_genes': 'rRNA16S genes', 'tRNA_genes': 'tRNA genes', 'gram_stain': 'Gram stain', 'sporulation': 'Sporulation', 'motility': 'Motility', 'range_salinity': 'Halophiles', 'facultative_respiration': 'Facultatives', 'anaerobic_respiration': 'Anaerobes', 'aerobic_respiration':'Aerobes' ,'mesophilic_range_tmp': 'Mesophiles', 'thermophilic_range_tmp':'Thermophiles', 'psychrophilic_range_tmp': 'Psychrophiles', 'bacillus_cell_shape': 'Bacillus', 'coccus_cell_shape': 'Coccus', 'filament_cell_shape': 'Filament', 'coccobacillus_cell_shape': 'Coccobacillus', 'vibrio_cell_shape': 'Vibrio', 'spiral_cell_shape': 'Spiral'}\n",
    "\n",
    "# Visualization\n",
    "figure = plt.figure(figsize=(3, 6))\n",
    "gs_master = GridSpec(nrows=2, ncols=2, height_ratios=[8, 10], width_ratios=[1.5, 2], hspace=0.55, wspace=0.25)\n",
    "\n",
    "### Loliplot: continuous traits\n",
    "gs_1 = GridSpecFromSubplotSpec(nrows=1, ncols=1, subplot_spec=gs_master[0, 0])\n",
    "axes_1 = figure.add_subplot(gs_1[:, :])\n",
    "\n",
    "height = list(reversed(list(phylod_empo_nt['threshold'])))\n",
    "markerline, _, _ = axes_1.stem(height, linefmt='k-', markerfmt='ko', orientation='horizontal', basefmt='black')\n",
    "plt.setp(markerline, markersize = 4)\n",
    "for i, t in enumerate(height):\n",
    "    axes_1.text(t+0.2, i, f'{t: .2f}', fontsize=7,\n",
    "                horizontalalignment='left', verticalalignment='center_baseline')\n",
    "\n",
    "axes_1.set_xlabel('Thresholds for \\nPhylogenetic distance')\n",
    "axes_1.set_yticks(np.arange(phylod_empo_nt.shape[0]))\n",
    "axes_1.set_yticklabels([titles[t] for t in list(reversed(list(phylod_empo_nt.index)))])\n",
    "axes_1.set_title('8 continuous traits', fontsize=10)\n",
    "\n",
    "# threshold_phylodistance\n",
    "axes_1.set_xlim(0, 3)\n",
    "\n",
    "axes_1.xaxis.set_major_locator(FixedLocator(np.arange(start=0, stop=2+1, step=1)))\n",
    "axes_1.xaxis.set_major_formatter(FormatStrFormatter('%.1f'))\n",
    "axes_1.xaxis.set_minor_locator(FixedLocator(np.arange(start=0, stop=2+0.25, step=0.25)))\n",
    "axes_1.grid(axis='x', alpha=0.1, color='grey')\n",
    "\n",
    "axes_1.spines['top'].set(color='grey', linewidth=.15)\n",
    "axes_1.spines['right'].set(color='grey', linewidth=.15)\n",
    "\n",
    "### Heatplot: continuous traits\n",
    "gs_2 = GridSpecFromSubplotSpec(nrows=1, ncols=1, subplot_spec=gs_master[0, 1])\n",
    "axes_2 = figure.add_subplot(gs_2[:, :])\n",
    "\n",
    "sns.heatmap(phylod_empo_nt.drop(['threshold'], axis='columns'), cmap='magma', cbar=False,\n",
    "            square=True, vmax=1, vmin=0, annot=True, annot_kws={\"size\":6}, fmt='.2f', xticklabels=1, yticklabels=1, linewidths=.5, ax=axes_2)\n",
    "\n",
    "axes_2.set_xticklabels(phylod_empo_nt.columns[1:], rotation=90)\n",
    "axes_2.set_yticklabels([])\n",
    "axes_2.set_ylabel('')\n",
    "axes_2.tick_params(bottom=False, left=False)\n",
    "axes_2.tick_params(axis='x', which='major', pad=0)\n",
    "\n",
    "cbar = figure.colorbar(axes_2.collections[0], shrink=0.7, aspect=30, anchor=(1.0, 0.0), pad=-0.05)\n",
    "cbar.set_label('Relative abundance')\n",
    "cbar.ax.tick_params(labelsize=7)\n",
    "cbar.set_ticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "cbar.outline.set_linewidth(0.5)\n",
    "\n",
    "### Loliplot: categorical traits\n",
    "gs_3 = GridSpecFromSubplotSpec(nrows=1, ncols=1, subplot_spec=gs_master[1, 0])\n",
    "axes_3 = figure.add_subplot(gs_3[:, :])\n",
    "\n",
    "height = list(reversed(list(phylod_empo_ct['threshold'])))\n",
    "markerline, _, _ = axes_3.stem(height, linefmt='k-', markerfmt='ko', orientation='horizontal', basefmt='black')\n",
    "plt.setp(markerline, markersize = 4)\n",
    "for i, t in enumerate(height):\n",
    "    axes_3.text(t+0.2, i, f'{t: .2f}', fontsize=7, horizontalalignment='left', verticalalignment='center_baseline')\n",
    "\n",
    "axes_3.set_xlabel('Thresholds for \\nPhylogenetic distance')\n",
    "axes_3.set_yticks(np.arange(phylod_empo_ct.shape[0]))\n",
    "axes_3.set_yticklabels([titles[t] for t in list(reversed(list(phylod_empo_ct.index)))])\n",
    "axes_3.set_title('10 categorical traits', fontsize=10)\n",
    "\n",
    "# threshold_phylodistance\n",
    "axes_3.set_xlim(0, 3.0)\n",
    "\n",
    "axes_3.xaxis.set_major_locator(FixedLocator(np.arange(start=0, stop=2+1, step=1)))\n",
    "axes_3.xaxis.set_major_formatter(FormatStrFormatter('%.1f'))\n",
    "axes_3.xaxis.set_minor_locator(FixedLocator(np.arange(start=0, stop=2+0.25, step=0.25)))\n",
    "axes_3.grid(axis='x', alpha=0.1, color='grey')\n",
    "\n",
    "axes_3.spines['top'].set(color='grey', linewidth=.15)\n",
    "axes_3.spines['right'].set(color='grey', linewidth=.15)\n",
    "\n",
    "### Heatplot: continuous traits\n",
    "gs_4 = GridSpecFromSubplotSpec(nrows=1, ncols=1, subplot_spec=gs_master[1, 1])\n",
    "axes_4 = figure.add_subplot(gs_4[:, :])\n",
    "\n",
    "sns.heatmap(phylod_empo_ct.drop(['threshold'], axis='columns'), cmap='magma', cbar=False,\n",
    "            square=True, vmax=1, vmin=0, annot=True, annot_kws={\"size\":6}, fmt='.2f', xticklabels=1, yticklabels=1, linewidths=.5, ax=axes_4)\n",
    "\n",
    "axes_4.set_xticklabels(phylod_empo_ct.columns[1:], rotation=90)\n",
    "axes_4.tick_params(axis='x', which='major', pad=0)\n",
    "axes_4.set_yticklabels([])\n",
    "axes_4.set_ylabel('')\n",
    "axes_4.tick_params(bottom=False, left=False)\n",
    "\n",
    "cbar = figure.colorbar(axes_4.collections[0], shrink=0.65, aspect=30, anchor=(1.0, 0.0), pad=-0.05)\n",
    "cbar.set_label('Relative abundance')\n",
    "cbar.ax.tick_params(labelsize=7)\n",
    "cbar.set_ticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "cbar.outline.set_linewidth(0.5)\n",
    "\n",
    "plt.savefig(\"../../results/06_trait_autocorrelations/fig3a-3b.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bac2feature",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
