{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "# Application to 16S rRNA amplicon analysis of the Finnish children"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Environment settings\n",
    "```sh\n",
    "# Working Directory\n",
    "cd Bac2fFeature/scripts/07_application_gutmicrobiome\n",
    "# Output Directory\n",
    "directories=(\n",
    "    \"../../data/t1d\"\n",
    "    \"../../data/t1d/qiime2\"\n",
    "    \"../../data/t1d_samples\"\n",
    ")\n",
    "for dir in \"${directories[@]}\"; do\n",
    "  if [ ! -d \"$dir\" ]; then\n",
    "    mkdir -p \"$dir\"\n",
    "  fi\n",
    "done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DADA2 using Qiime2 pipeline\n",
    "Notion:  \n",
    "The number of forward reads and the number of reverse reads in some samples were different and using all samples resulted in a DADA2 error.  \n",
    "We therefore used 654 samples where the number of forward reads and the number of reverse reads were equal.\n",
    "```sh\n",
    "conda activate qiime2-2023.5\n",
    "# Import (Please replace <path_to_data_dir> in Manifest_study_sample.csv to bsolute path of data directory)\n",
    "qiime tools import --type 'SampleData[PairedEndSequencesWithQuality]' --input-path Manifest_study_sample.csv --output-path ../../data/t1d/demux_study_sample.qza --input-format PairedEndFastqManifestPhred33\n",
    "qiime demux summarize --i-data ../../data/t1d/qiime2/demux_study_sample.qza --o-visualization ../../data/t1d/qiime2/demux_study_sample.qzv\n",
    "# DADA2\n",
    "qiime dada2 denoise-paired --i-demultiplexed-seqs ../../data/t1d/qiime2/demux_study_sample.qza --p-trunc-len-f 0 --p-trunc-len-r 160 --p-n-threads 4 --o-table ../../data/t1d/qiime2/table_study_sample.qza --o-denoising-stats ../../data/t1d/qiime2/stats_study_sample.qza --o-representative-sequences ../../data/t1d/qiime2/reps_sequences_study_sample.qza --verbose\n",
    "# Export\n",
    "qiime tools export --input-path ../../data/t1d/qiime2/reps_sequences_study_sample.qza --output-path ../../data/t1d/qiime2/\n",
    "qiime tools export --input-path ../../data/t1d/qiime2/table_study_sample.qza --output-path ../../data/t1d/qiime2/\n",
    "biom convert -i ../../data/t1d/qiime2/feature-table.biom -o ../../data/t1d/qiime2/feature_table_study_sample.tsv --to-tsv\n",
    "qiime tools export --input-path ../../data/t1d/qiime2/taxonomy_study_sample.qza --output-path ../../data/t1d/qiime2/\n",
    "```\n",
    "\n",
    "```sh\n",
    "wget -P ../../data/t1d/qiime2 https://data.qiime2.org/2023.7/common/silva-138-99-nb-classifier.qza\n",
    "# Classifying taxonomy\n",
    "qiime feature-classifier classify-sklearn --i-classifier ../../data/t1d/qiime2/silva-138-99-nb-classifier.qza --i-reads ../../data/t1d/qiime2/reps_sequences_study_sample.qza --o-classification ../../data/t1d/qiime2/taxonomy_study_sample.qza\n",
    "# Barplot\n",
    "qiime taxa barplot --i-table ../../data/t1d/qiime2/table_study_sample.qza --i-taxonomy ../../data/t1d/qiime2/taxonomy_study_sample.qza --m-metadata-file ../../data/t1d/qiime2/qiime_metadata.tsv --o-visualization ../../data/t1d/qiime2/taxa_barplot.qzv\n",
    "# Please obtain file \"level-7.csv\" from Qiime2 viewer and copy it to ../../data/t1d/qiime2/level-7.csv\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import math\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "matplotlib.rcParams['font.family']       = 'Arial'\n",
    "matplotlib.rcParams['font.sans-serif']   = [\"Arial\",\"DejaVu Sans\",\"Lucida Grande\",\"Verdana\"]\n",
    "matplotlib.rcParams[\"figure.figsize\"]    = [4,3]\n",
    "matplotlib.rcParams[\"font.size\"]         = 10\n",
    "matplotlib.rcParams[\"axes.labelcolor\"]   = \"#000000\"\n",
    "matplotlib.rcParams[\"axes.linewidth\"]    = 1.0\n",
    "matplotlib.rcParams[\"xtick.major.width\"] = 1.0\n",
    "matplotlib.rcParams[\"ytick.major.width\"] = 1.0\n",
    "cmap1 = plt.cm.tab20\n",
    "cmap2 = plt.cm.Set3\n",
    "\n",
    "from scipy.stats import pearsonr\n",
    "from matplotlib.ticker import MultipleLocator, FormatStrFormatter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxa = pd.read_csv('../../data/t1d/qiime2/level-7.csv', sep=',')\n",
    "\n",
    "# Normarize\n",
    "taxa_cols = [c for c in taxa.columns if 'd__' in c]\n",
    "taxa[taxa_cols] = taxa[taxa_cols] / taxa[taxa_cols].sum(axis=1)[:,None]\n",
    "\n",
    "# Melt\n",
    "taxa_melt = taxa.melt(id_vars=['index'], value_vars=taxa_cols, var_name='taxa', value_name='abundance')\n",
    "\n",
    "# Split taxonomy columns\n",
    "taxa_split = taxa_melt['taxa'].str.split(';', expand=True)\n",
    "taxa_split = taxa_split.set_axis(['domain', 'phylum', 'class', 'order', 'family', 'genus', 'species'],\n",
    "                                  axis=1, copy=False)\n",
    "\n",
    "taxa_melt_split = pd.concat([taxa_melt, taxa_split], axis=1)\n",
    "\n",
    "tax_level = 'family'\n",
    "tax_num = 7\n",
    "\n",
    "# Top major tax\n",
    "tax_sum = taxa_melt_split[[tax_level, 'abundance']].groupby(tax_level).sum()\n",
    "tax_sum = tax_sum.reset_index().sort_values(by='abundance', ascending=False)\n",
    "top_tax = list(tax_sum[tax_level].head(tax_num))\n",
    "\n",
    "taxa_melt_split['major_tax'] = taxa_melt_split[tax_level].apply(lambda x: x if x in top_tax else 'others')\n",
    "\n",
    "# Groupby with tax\n",
    "tax_level_abd = taxa_melt_split.groupby(['index', 'major_tax']).sum()\n",
    "tax_level_abd.reset_index(inplace=True)\n",
    "tax_level_abd.sort_values(by='index').reset_index(inplace=True)\n",
    "\n",
    "tax_level_abd_age = pd.merge(tax_level_abd, taxa[['index', 'Age_at_Collection']], how='left', on='index')\n",
    "tax_level_abd_age.sort_values(by=['Age_at_Collection', 'index'], inplace=True)\n",
    "tax_level_abd_age.reset_index(inplace=True, drop=True)\n",
    "\n",
    "# Binning the age at collection\n",
    "bin_width=100\n",
    "bin_max = tax_level_abd_age['Age_at_Collection'].max() + 1\n",
    "bin_age = list(np.arange(0, tax_level_abd_age['Age_at_Collection'].max(), 100)) + [bin_max]\n",
    "tax_level_abd_age['age_category'] = pd.cut(tax_level_abd_age['Age_at_Collection'], bin_age, right=False)\n",
    "\n",
    "tax_age = tax_level_abd_age[['age_category', 'major_tax', 'abundance']].groupby(['age_category', 'major_tax']).mean()\n",
    "tax_age.reset_index(inplace=True)\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(figsize=(4, 2))\n",
    "\n",
    "age_list = list(tax_age['age_category'].unique())\n",
    "tax_list = list(tax_age['major_tax'].unique())\n",
    "\n",
    "xs = bin_age\n",
    "for x, c in zip(xs, age_list):\n",
    "    tax_age_cat = tax_age[tax_age['age_category']==c]\n",
    "    for i, p in enumerate(tax_list):\n",
    "        abd = np.array(tax_age_cat[tax_age_cat['major_tax']==p]['abundance'])\n",
    "        if i == 0:\n",
    "            ax.bar(x, abd, width=90, color=cmap2.colors[i], align='edge')\n",
    "            sum = abd\n",
    "        elif p == 'others':\n",
    "            ax.bar(x, abd, bottom=sum, width=90, color='grey', align='edge')\n",
    "            sum += abd\n",
    "        else:\n",
    "            ax.bar(x, abd, bottom=sum, width=90, color=cmap2.colors[i], align='edge')\n",
    "            sum += abd\n",
    "\n",
    "ax.legend([t.replace('f__', '') for t in tax_list],\n",
    "          bbox_to_anchor=(1.03, -0.05), loc='lower left', borderaxespad=0, ncol=1, frameon=False)\n",
    "\n",
    "ax.set_ylim(0, 1)\n",
    "ax.xaxis.set_major_locator(MultipleLocator(200))\n",
    "ax.xaxis.set_minor_locator(MultipleLocator(100))\n",
    "\n",
    "ax.set_xlabel('Age at Collection (days)')\n",
    "ax.set_ylabel('Relative abundance')\n",
    "\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "\n",
    "plt.savefig(\"../../results/07_application_gutmicrobiome/fig4a.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trait prediction using Bac2Feature\n",
    "```sh\n",
    "conda activate bac2feature_experiment\n",
    "bac2feature -s ../../data/t1d/qiime2/dna-sequences.fasta -o ../../data/t1d/predicted_trait.tsv -m phylogeny --ref_dir ../../data/ref_bac2feature/phylogeny --ref_trait ../../data/ref_bac2feature/trait_bac2feature.tsv --intermediate_dir ../../data/t1d/intermediate_dir --calculate_NSTI\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "predicted_trait_path = \"../../data/t1d/predicted_trait.tsv\"\n",
    "pred = pd.read_csv(predicted_trait_path, sep=\"\\t\")\n",
    "\n",
    "threshold_phylodistance = pd.read_csv(\"../../data/trait_autocorrelations/threshold_phylodistance.tsv\", sep= \"\\t\", index_col=0)\n",
    "threshold_phylodistance.rename(columns={\"cor_0.5\":\"threshold\"}, inplace=True)\n",
    "\n",
    "# Feature table\n",
    "table = pd.read_csv(\"../../data/t1d/qiime2/feature_table_study_sample.tsv\", sep=\"\\t\", header=1)\n",
    "# rename the first column\n",
    "table.rename(columns={\"#OTU ID\": \"sequence\"}, inplace=True)\n",
    "\n",
    "# Metadata from the paper\n",
    "metadata = pd.read_csv(\"../../data/t1d/qiime2/metadata_study_sample.tsv\", sep=\"\\t\", dtype=str)\n",
    "\n",
    "taxonomy = pd.read_csv('../../data/t1d/qiime2/taxonomy.tsv', sep='\\t')\n",
    "# Split taxonomy columns\n",
    "taxonomy_split = taxonomy['Taxon'].str.split(';', expand=True)\n",
    "taxonomy_split = taxonomy_split.set_axis(['domain', 'phylum', 'class', 'order', 'family', 'genus', 'species'],\n",
    "                                         axis=1, copy=False)\n",
    "\n",
    "taxonomy_split = pd.concat([taxonomy['Feature ID'], taxonomy_split], axis=1)\n",
    "# rename the first column\n",
    "taxonomy_split.rename(columns={\"Feature ID\": \"sequence\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def normalize_by_phylum(table, taxonomy_split, p=None): # Limiting within phylum\n",
    "    abd_cols = table.columns[1:]\n",
    "    table_tax = pd.merge(table, taxonomy_split, how='inner', on='sequence')\n",
    "    if p is not None:\n",
    "        table_tax = table_tax[table_tax['phylum']==p]\n",
    "    table_tax.reset_index(drop=True, inplace=True)\n",
    "    table_tax[abd_cols] = table_tax[abd_cols] / table_tax[abd_cols].sum(axis=0)\n",
    "    return table_tax[table.columns]\n",
    "\n",
    "titles = {'cell_diameter': 'Cell diameter', 'cell_length': 'Cell length', 'doubling_h': 'Doubling time', 'growth_tmp': 'Growth temp.', 'optimum_tmp': 'Optimum temp.', 'optimum_ph': 'Optimum pH', 'genome_size': 'Genome size', 'gc_content': 'GC content', 'coding_genes': 'Coding genes', 'rRNA16S_genes': 'rRNA16S genes', 'tRNA_genes': 'tRNA genes', 'gram_stain': 'Gram stain', 'sporulation': 'Sporulation', 'motility': 'Motility', 'range_salinity': 'Halophiles', 'facultative_respiration': 'Facultatives', 'anaerobic_respiration': 'Anaerobes', 'aerobic_respiration':'Aerobes' ,'mesophilic_range_tmp': 'Mesophiles', 'thermophilic_range_tmp':'Thermophiles', 'psychrophilic_range_tmp': 'Psychrophiles', 'bacillus_cell_shape': 'Bacillus', 'coccus_cell_shape': 'Coccus', 'filament_cell_shape': 'Filament', 'coccobacillus_cell_shape': 'Coccobacillus', 'vibrio_cell_shape': 'Vibrio', 'spiral_cell_shape': 'Spiral'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Continuous traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nt = [\"doubling_h\", \"genome_size\", \"gc_content\", \"coding_genes\", \"optimum_tmp\", \"growth_tmp\", \"rRNA16S_genes\", \"tRNA_genes\"]\n",
    "\n",
    "weighted_nt = dict()\n",
    "is_first = True\n",
    "\n",
    "p = None # Not limiting within phylum\n",
    "\n",
    "# Calculate community weighted mean\n",
    "for t in nt:\n",
    "    # filtering\n",
    "    x = pd.merge(table, pred[[\"sequence\", t+\"_nsti\"]], how=\"inner\", on=\"sequence\")\n",
    "    filtered_table = x[x[t+\"_nsti\"] < threshold_phylodistance.loc[t, \"threshold\"]].drop(t+\"_nsti\", axis=1).copy()\n",
    "\n",
    "    # normalize\n",
    "    filtered_table = normalize_by_phylum(filtered_table, taxonomy_split, p)\n",
    "    # melt table\n",
    "    melt_table = filtered_table.melt(id_vars=\"sequence\", value_name=\"abundance\", var_name=\"sample\")\n",
    "    # Merging tooks several seconds\n",
    "    melt_trait = pd.merge(melt_table, pred[[\"sequence\", t]], how=\"inner\", on=\"sequence\")\n",
    "    melt_trait[\"weighted_nt\"] = melt_trait[\"abundance\"] * melt_trait[t]\n",
    "\n",
    "    cwm_nt = melt_trait[[\"sample\", \"weighted_nt\"]].groupby(by=\"sample\").sum()\n",
    "    cwm_nt = cwm_nt.reset_index()\n",
    "    cwm_nt = cwm_nt.sort_values(by=\"sample\")\n",
    "\n",
    "    if len(list(cwm_nt[\"weighted_nt\"])) == 0:\n",
    "        continue\n",
    "    if is_first:\n",
    "        is_first = False\n",
    "        weighted_nt[\"sample\"] = list(cwm_nt[\"sample\"])\n",
    "    weighted_nt[t] = list(cwm_nt[\"weighted_nt\"])\n",
    "\n",
    "cwn_nt_concat = pd.DataFrame(weighted_nt)\n",
    "cwm_nt_metadata = pd.merge(cwn_nt_concat, metadata, how=\"inner\", left_on=\"sample\", right_on=\"G_id\")\n",
    "\n",
    "cwm_nt_metadata[\"Age_at_Collection\"] = cwm_nt_metadata[\"Age_at_Collection\"].astype(int)\n",
    "\n",
    "bin_width = 100\n",
    "bin_max = 1200\n",
    "bin_list = list(np.arange(start=0, stop=bin_max+1, step=bin_width))\n",
    "bin_mid_list = list(np.arange(start=bin_width/2, stop=bin_max+1, step=bin_width))\n",
    "cwm_nt_metadata['cut'] = pd.cut(cwm_nt_metadata['Age_at_Collection'], bins=bin_list)\n",
    "\n",
    "# Visualization\n",
    "matplotlib.rcParams['axes.xmargin'] = 0.05\n",
    "matplotlib.rcParams['axes.ymargin'] = 0.05\n",
    "\n",
    "trait_list = ['doubling_h', 'optimum_tmp', 'genome_size', 'gc_content', 'rRNA16S_genes', 'tRNA_genes']\n",
    "yunit_dict = {'doubling_h': '(log[h])', 'optimum_tmp': '(Â°C)',\n",
    "              'genome_size': '(Mb)', 'gc_content': '(%)', 'rRNA16S_genes': '(#)', 'tRNA_genes': '(#)'}\n",
    "\n",
    "col = 2\n",
    "row = math.ceil(len(trait_list) / col)\n",
    "fig, axes = plt.subplots(row, col, figsize=(2*col, 2*row))\n",
    "\n",
    "for t, ax in zip(trait_list, axes.flatten()):\n",
    "    x = cwm_nt_metadata[\"Age_at_Collection\"]\n",
    "    y = cwm_nt_metadata[t]\n",
    "\n",
    "    if t == 'genome_size': # Convert to Mb\n",
    "        y = y / 1e+6\n",
    "\n",
    "    bins = 10\n",
    "    hist, xedges, yedges = np.histogram2d(x, y, bins=bins)\n",
    "\n",
    "    # Set Color based on density\n",
    "    hist = (hist - hist.min().min()) / hist.sum().sum()\n",
    "\n",
    "    x_indices = np.digitize(x, xedges) - 1\n",
    "    y_indices = np.digitize(y, yedges) - 1\n",
    "\n",
    "    x_indices[x_indices < 0] = 0\n",
    "    x_indices[x_indices >= bins] = bins - 1\n",
    "    y_indices[y_indices < 0] = 0\n",
    "    y_indices[y_indices >= bins] = bins - 1\n",
    "\n",
    "    colors = hist[x_indices, y_indices]\n",
    "\n",
    "    scatter = ax.scatter(x, y, c=colors, cmap='Oranges', s=20, edgecolors='grey', linewidths=0.1)\n",
    "\n",
    "    ymin, ymax = ax.get_ylim()\n",
    "    ax.set_ylim(ymin, ymax + (ymax - ymin) * 0.20)\n",
    "\n",
    "    ax.text(0.05, 0.95, titles[t], transform=ax.transAxes, verticalalignment='top')\n",
    "    ax.xaxis.set_major_locator(MultipleLocator(300))\n",
    "    ax.tick_params('x', pad = 3)\n",
    "    ax.tick_params('y', pad = 1)\n",
    "\n",
    "    ax.text(-0.01, 0.97, yunit_dict[t], ha='right', transform=ax.transAxes)\n",
    "\n",
    "plt.savefig(\"../../results/07_application_gutmicrobiome/fig4b.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Categorical traits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct = ['gram_stain', 'sporulation', 'motility', 'anaerobic_respiration', 'mesophilic_range_tmp', 'thermophilic_range_tmp', 'bacillus_cell_shape', 'coccus_cell_shape', 'filament_cell_shape', 'spiral_cell_shape']\n",
    "\n",
    "weighted_ct = dict()\n",
    "is_first = True\n",
    "\n",
    "p = None # Not limiting within phylum\n",
    "\n",
    "for t in ct:\n",
    "    x = pd.merge(table, pred[[\"sequence\", t+\"_nsti\"]], how=\"inner\", on=\"sequence\")\n",
    "    filtered_table = x[x[t+\"_nsti\"] < threshold_phylodistance.loc[t, \"threshold\"]].drop(t+\"_nsti\", axis=1).copy()\n",
    "\n",
    "    # normalize\n",
    "    filtered_table = normalize_by_phylum(filtered_table, taxonomy_split, p)\n",
    "    # melt table\n",
    "    melt_table = filtered_table.melt(id_vars=\"sequence\", value_name=\"abundance\", var_name=\"sample\")\n",
    "    # Merge\n",
    "    melt_trait = pd.merge(melt_table, pred[[\"sequence\", t]], how=\"inner\", on=\"sequence\")\n",
    "\n",
    "    melt_trait[\"weighted_ct\"] = melt_trait[\"abundance\"] * melt_trait[t]\n",
    "\n",
    "    cwm_ct = melt_trait[[\"sample\", \"weighted_ct\"]].groupby(by=\"sample\").sum()\n",
    "    cwm_ct = cwm_ct.reset_index()\n",
    "    cwm_ct = cwm_ct.sort_values(by=\"sample\")\n",
    "    if len(list(cwm_ct[\"weighted_ct\"])) == 0:\n",
    "        continue\n",
    "    if is_first:\n",
    "        is_first = False\n",
    "        weighted_ct[\"sample\"] = list(cwm_ct[\"sample\"])\n",
    "    weighted_ct[t] = list(cwm_ct[\"weighted_ct\"])\n",
    "\n",
    "cwn_ct_concat = pd.DataFrame(weighted_ct)\n",
    "cwm_ct_metadata = pd.merge(cwn_ct_concat, metadata, how=\"inner\", left_on=\"sample\", right_on=\"G_id\")\n",
    "cwm_ct_metadata[\"Age_at_Collection\"] = cwm_ct_metadata[\"Age_at_Collection\"].astype(int)\n",
    "\n",
    "bin_width=100\n",
    "bin_max = 1200\n",
    "bin_list = list(np.arange(start=0, stop=bin_max+1, step=bin_width))\n",
    "bin_mid_list = list(np.arange(start=bin_width/2, stop=bin_max+1, step=bin_width))\n",
    "cwm_ct_metadata['cut'] = pd.cut(cwm_ct_metadata['Age_at_Collection'], bins=bin_list)\n",
    "\n",
    "# Visualization\n",
    "matplotlib.rcParams['axes.xmargin'] = 0.05\n",
    "matplotlib.rcParams['axes.ymargin'] = 0.05\n",
    "\n",
    "trait_list = ['gram_stain', 'sporulation', 'motility', 'anaerobic_respiration', 'bacillus_cell_shape', 'coccus_cell_shape']\n",
    "\n",
    "col = 2\n",
    "row = math.ceil(len(trait_list) / col)\n",
    "fig, axes = plt.subplots(row, col, figsize=(2*col, 2*row))\n",
    "\n",
    "for t, ax in zip(trait_list, axes.flatten()):\n",
    "    x = cwm_ct_metadata[\"Age_at_Collection\"]\n",
    "    y = cwm_ct_metadata[t] * 100\n",
    "\n",
    "    bins = 10\n",
    "    hist, xedges, yedges = np.histogram2d(x, y, bins=bins)\n",
    "\n",
    "    # Set Color based on density\n",
    "    hist = (hist - hist.min().min()) / hist.sum().sum()\n",
    "\n",
    "    x_indices = np.digitize(x, xedges) - 1\n",
    "    y_indices = np.digitize(y, yedges) - 1\n",
    "\n",
    "    x_indices[x_indices < 0] = 0\n",
    "    x_indices[x_indices >= bins] = bins - 1\n",
    "    y_indices[y_indices < 0] = 0\n",
    "    y_indices[y_indices >= bins] = bins - 1\n",
    "\n",
    "    colors = hist[x_indices, y_indices]\n",
    "\n",
    "    scatter = ax.scatter(x, y, c=colors, cmap='Oranges', s=15, edgecolors='grey', linewidths=0.1)\n",
    "\n",
    "    ax.set_ylim(-5, 125)\n",
    "\n",
    "    ax.text(0.05, 0.95, titles[t], transform=ax.transAxes, verticalalignment='top')\n",
    "\n",
    "    ax.xaxis.set_major_locator(MultipleLocator(300))\n",
    "    ax.tick_params('x', pad = 3)\n",
    "    ax.tick_params('y', pad = 1)\n",
    "    ax.text(-0.01, 0.97, '(%)', ha='right', transform=ax.transAxes)\n",
    "\n",
    "plt.savefig(\"../../results/07_application_gutmicrobiome/fig4c.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Categorical traits within Firmicutes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ct = ['gram_stain', 'sporulation', 'motility', 'anaerobic_respiration', 'mesophilic_range_tmp', 'thermophilic_range_tmp', 'bacillus_cell_shape', 'coccus_cell_shape', 'filament_cell_shape', 'spiral_cell_shape']\n",
    "\n",
    "def above_abd_thres(df, focal_col, thres_col, thres):\n",
    "    res = df[focal_col]\n",
    "    if df[thres_col] < thres:\n",
    "        res = np.nan\n",
    "    return res\n",
    "\n",
    "p = ' p__Firmicutes' # Limiting within phylum Firmicutes\n",
    "\n",
    "weighted_ct = dict()\n",
    "is_first = True\n",
    "\n",
    "for t in ct:\n",
    "    x = pd.merge(table, pred[[\"sequence\", t+\"_nsti\"]], how=\"inner\", on=\"sequence\")\n",
    "    filtered_table = x[x[t+\"_nsti\"] < threshold_phylodistance.loc[t, \"threshold\"]].drop(t+\"_nsti\", axis=1).copy()\n",
    "\n",
    "    # normalize\n",
    "    filtered_table = normalize_by_phylum(filtered_table, taxonomy_split, p)\n",
    "    # melt table\n",
    "    melt_table = filtered_table.melt(id_vars=\"sequence\", value_name=\"abundance\", var_name=\"sample\")\n",
    "    # Merge\n",
    "    melt_trait = pd.merge(melt_table, pred[[\"sequence\", t]], how=\"inner\", on=\"sequence\")\n",
    "    melt_trait[\"weighted_ct\"] = melt_trait[\"abundance\"] * melt_trait[t]\n",
    "\n",
    "    cwm_ct = melt_trait[[\"sample\", \"abundance\", \"weighted_ct\"]].groupby(by=\"sample\").sum()\n",
    "    cwm_ct = cwm_ct.reset_index()\n",
    "    cwm_ct['weighted_ct'] = cwm_ct.apply(\n",
    "        lambda df: above_abd_thres(df, focal_col='weighted_ct', thres_col='abundance', thres=0.01), axis=1)\n",
    "    cwm_ct = cwm_ct.sort_values(by=\"sample\")\n",
    "    if len(list(cwm_ct[\"weighted_ct\"])) == 0:\n",
    "            continue\n",
    "    if is_first:\n",
    "        is_first = False\n",
    "        weighted_ct[\"sample\"] = list(cwm_ct[\"sample\"])\n",
    "    weighted_ct[t] = list(cwm_ct[\"weighted_ct\"])\n",
    "\n",
    "cwn_ct_concat = pd.DataFrame(weighted_ct)\n",
    "cwm_ct_metadata = pd.merge(cwn_ct_concat, metadata, how=\"inner\", left_on=\"sample\", right_on=\"G_id\")\n",
    "\n",
    "cwm_ct_metadata[\"Age_at_Collection\"] = cwm_ct_metadata[\"Age_at_Collection\"].astype(int)\n",
    "bin_width=100\n",
    "bin_max = 1200\n",
    "bin_list = list(np.arange(start=0, stop=bin_max+1, step=bin_width))\n",
    "bin_mid_list = list(np.arange(start=bin_width/2, stop=bin_max+1, step=bin_width))\n",
    "cwm_ct_metadata['cut'] = pd.cut(cwm_ct_metadata['Age_at_Collection'], bins=bin_list)\n",
    "\n",
    "# Visualization\n",
    "matplotlib.rcParams['axes.xmargin'] = 0.05\n",
    "matplotlib.rcParams['axes.ymargin'] = 0.05\n",
    "\n",
    "trait_list = ['bacillus_cell_shape', 'coccus_cell_shape']\n",
    "\n",
    "col = 2\n",
    "row = math.ceil(len(trait_list) / col)\n",
    "fig, axes = plt.subplots(row, col, figsize=(2*col, 2*row))\n",
    "\n",
    "for t, ax in zip(trait_list, axes.flatten()):\n",
    "    x = cwm_ct_metadata[\"Age_at_Collection\"]\n",
    "    y = cwm_ct_metadata[t] * 100\n",
    "\n",
    "    bins = 10\n",
    "    hist, xedges, yedges = np.histogram2d(x, y, bins=bins)\n",
    "\n",
    "    # Set Color based on density\n",
    "    hist = (hist - hist.min().min()) / hist.sum().sum()\n",
    "\n",
    "    x_indices = np.digitize(x, xedges) - 1\n",
    "    y_indices = np.digitize(y, yedges) - 1\n",
    "\n",
    "    x_indices[x_indices < 0] = 0\n",
    "    x_indices[x_indices >= bins] = bins - 1\n",
    "    y_indices[y_indices < 0] = 0\n",
    "    y_indices[y_indices >= bins] = bins - 1\n",
    "\n",
    "    colors = hist[x_indices, y_indices]\n",
    "\n",
    "    scatter = ax.scatter(x, y, c=colors, cmap='Oranges', s=20, edgecolors='grey', linewidths=0.1)\n",
    "\n",
    "    ax.set_ylim(-5, 125)\n",
    "    ax.text(0.05, 0.95, titles[t], transform=ax.transAxes, verticalalignment='top')\n",
    "    ax.xaxis.set_major_locator(MultipleLocator(300))\n",
    "    ax.tick_params('x', pad = 3)\n",
    "    ax.tick_params('y', pad = 1)\n",
    "    ax.text(-0.01, 0.97, '(%)', ha='right', transform=ax.transAxes)\n",
    "\n",
    "plt.savefig(\"../../results/07_application_gutmicrobiome/fig4d.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bac2feature",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
