{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28e2c7f6",
   "metadata": {},
   "source": [
    "# Application to 16S rRNA amplicon analysis of the lake and soil microbiome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca5b05c1",
   "metadata": {},
   "source": [
    "## Environment settings\n",
    "```sh\n",
    "# Working Directory\n",
    "cd Bac2fFeature/scripts/07_application_gutmicrobiome\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "386540ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import os\n",
    "import math\n",
    "\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import spearmanr\n",
    "from sklearn.linear_model import LinearRegression\n",
    "import statsmodels.api as sm\n",
    "\n",
    "matplotlib.rcParams['font.family']       = 'Arial'\n",
    "matplotlib.rcParams['font.sans-serif']   = [\"Arial\",\"DejaVu Sans\",\"Lucida Grande\",\"Verdana\"]\n",
    "matplotlib.rcParams['figure.figsize']    = [4,3]\n",
    "matplotlib.rcParams['font.size']         = 10\n",
    "matplotlib.rcParams[\"axes.labelcolor\"]   = \"#000000\"\n",
    "matplotlib.rcParams[\"axes.linewidth\"]    = 1.0\n",
    "matplotlib.rcParams[\"xtick.major.width\"] = 1.0\n",
    "matplotlib.rcParams[\"ytick.major.width\"] = 1.0\n",
    "cmap1 = plt.cm.tab20\n",
    "cmap2 = plt.cm.Set3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a0eae43",
   "metadata": {},
   "source": [
    "## Lake microbiome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50694246",
   "metadata": {},
   "source": [
    "### DADA2 parameters\n",
    "```sh\n",
    "qiime dada2 denoise-paired --i-demultiplexed-seqs ../../data/2023-10-05/demux_cutadapt_SRP136143.qza --p-trunc-len-f 240 --p-trunc-len-r 240 --p-min-overlap 20  --o-representative-sequences ../../data/2023-10-05/reps_seqs_cutadapt_SRP136143.qza --o-table ../../data/2023-10-05/table_cutadapt_SRP136143.qza --o-denoising-stats ../../data/2023-10-05/denoising_stats_cutadapt_SRP136143.qza --p-n-threads 40\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36778b1e",
   "metadata": {},
   "source": [
    "### Trait prediction\n",
    "```sh\n",
    "bac2feature -s asv_seqs_PRJEB27579.fasta -o predicted_trait_PRJEB27579.tsv -m phylogeny --ref_dir ../../data/ref_bac2feature/phylogeny/ --ref_table ../../data/ref_bac2feature.tsv --intermediate_dir intermediate_dir --threads 20 --calculate_NSTI\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1bb989ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "trait = pd.read_csv(\"predicted_trait_PRJEB27579.tsv\", sep=\"\\t\")\n",
    "feature_table = pd.read_csv(\"table_PRJEB27579.tsv\", sep=\"\\t\", header=1)\n",
    "metadata_path = \"metadata_PRJEB27579.tsv\"\n",
    "metadata = pd.read_csv(metadata_path, sep=\"\\t\")\n",
    "metadata = metadata.rename(columns={\"sampleid\": \"sample\"})\n",
    "\n",
    "# normalize\n",
    "abundance = feature_table.iloc[:, 1:]\n",
    "feature_table.iloc[:, 1:] = abundance / abundance.sum(axis=0)\n",
    "# rename id column\n",
    "feature_table = feature_table.rename(columns={\"#OTU ID\": \"sequence\"})\n",
    "\n",
    "feature_table = feature_table.melt(id_vars=\"sequence\",\n",
    "                                   value_name=\"abundance\",\n",
    "                                   var_name=\"sample\"\n",
    "                )\n",
    "\n",
    "# Merge feature table with trait\n",
    "feature_trait = pd.merge(feature_table, trait, how=\"inner\", on=\"sequence\")\n",
    "\n",
    "threshold_phylodistance = pd.read_csv(\"../../data/trait_autocorrelations/threshold_phylodistance.tsv\", sep= \"\\t\", index_col=0)\n",
    "threshold_phylodistance.rename(columns={\"cor_0.5\": \"threshold\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8ea3fd91",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Focus on anaerobic respiration\n",
    "t = \"anaerobic_respiration\"\n",
    "\n",
    "def rm_spurious_prediction(df, threshold_phylodistance, t):\n",
    "    threshold = threshold_phylodistance.loc[t, \"threshold\"]\n",
    "    # remove\n",
    "    if df[t+\"_nsti\"] > threshold:\n",
    "        return np.nan\n",
    "    else:\n",
    "        return df[\"abundance\"]\n",
    "\n",
    "feature_trait[\"fil_abd\"] = feature_trait.apply(\n",
    "    lambda x: rm_spurious_prediction(x, threshold_phylodistance, t), axis=1)\n",
    "\n",
    "feature_trait[t+\"_fil_abd\"] = feature_trait[\"fil_abd\"] * feature_trait[t]\n",
    "\n",
    "# Nomalization\n",
    "ratio = feature_trait.groupby(\"sample\", as_index=False)[\"fil_abd\"].sum()\n",
    "ratio[\"inv\"] = 1 / ratio[\"fil_abd\"]\n",
    "anaerobic_abd = feature_trait.groupby(\"sample\", as_index=False)[t+\"_fil_abd\"].sum()\n",
    "\n",
    "anaerobic_abd_normed = pd.merge(anaerobic_abd, ratio, how=\"inner\", on=\"sample\")\n",
    "anaerobic_abd_normed[t+\"_fil_abd_normed\"] = anaerobic_abd_normed[t+\"_fil_abd\"] * anaerobic_abd_normed[\"inv\"]\n",
    "\n",
    "anaerobic_abd_meta = pd.merge(anaerobic_abd_normed, metadata, how=\"inner\", on=\"sample\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e83bf18",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 1, figsize=(3, 2.5))\n",
    "x = anaerobic_abd_meta[\"Oxygen\"]\n",
    "y = 1 - anaerobic_abd_meta[t+\"_fil_abd_normed\"]\n",
    "\n",
    "for g, m, c in zip([\"Greifensee\", \"Lake Zug\", \"Lake Lugano\", \"Rotsee\"], [\"D\", \"s\", \"^\", \"o\"], [\"tab:blue\", \"tab:orange\", \"tab:green\", \"tab:red\"]):\n",
    "    x = anaerobic_abd_meta[anaerobic_abd_meta[\"Group\"]==g][\"Oxygen\"]\n",
    "    y = 1 - anaerobic_abd_meta[anaerobic_abd_meta[\"Group\"]==g][t+\"_fil_abd_normed\"]\n",
    "    sc = ax.scatter(x+1, y, marker=m, s=40, alpha=1, edgecolor=c, facecolor=\"None\")\n",
    "\n",
    "ax.set_xlabel(\"Oxygen (Âµmol/L)\", fontsize=14)\n",
    "ax.set_ylabel(\"Fraction of\\n oxygen tolerant spp.\", fontsize=14)\n",
    "\n",
    "ax.set_title(\"Lake water\", fontsize=14)\n",
    "ax.legend([\"Greifensee\", \"Lake Zug\", \"Lake Lugano\", \"Rotsee\"])\n",
    "\n",
    "spearman_corr, p_value = spearmanr(anaerobic_abd_meta[\"Oxygen\"], 1 - anaerobic_abd_meta[t+\"_fil_abd_normed\"], nan_policy='omit')\n",
    "print(f\"Spearman correlation coefficient: {spearman_corr:.3f}, p-value: {p_value:.2e}\")\n",
    "\n",
    "plt.savefig(\"../../results/08_application_soil_lake_microbiome/fig5a.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba4116d5",
   "metadata": {},
   "source": [
    "## Soil microbiome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bb3631e",
   "metadata": {},
   "source": [
    "### DADA2 parameters\n",
    "```sh\n",
    "qiime dada2 denoise-paired --i-demultiplexed-seqs ../../data/2023-10-05/demux_cutadapt_SRP136143.qza --p-trunc-len-f 240 --p-trunc-len-r 240 --p-min-overlap 20  --o-representative-sequences ../../data/2023-10-05/reps_seqs_cutadapt_SRP136143.qza --o-table ../../data/2023-10-05/table_cutadapt_SRP136143.qza --o-denoising-stats ../../data/2023-10-05/denoising_stats_cutadapt_SRP136143.qza --p-n-threads 40\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f14987a8",
   "metadata": {},
   "source": [
    "### Trait prediction\n",
    "```sh\n",
    "bac2feature -s asv_seqs_SRP136143.fasta -o predicted_trait_SRP136143.tsv -m phylogeny --ref_dir ../../data/ref_bac2feature/phylogeny/ --ref_table ../../data/ref_bac2feature.tsv --intermediate_dir intermediate_dir --threads 20 --calculate_NSTI\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b0d1555",
   "metadata": {},
   "outputs": [],
   "source": [
    "predicted_trait_path = \"predicted_trait_SRP136143.tsv\"\n",
    "pred = pd.read_csv(predicted_trait_path, sep=\"\\t\")\n",
    "\n",
    "# Feature table\n",
    "table = pd.read_csv(\"table_SRP136143.tsv\", sep=\"\\t\", header=1)\n",
    "# normalize\n",
    "abundance = table.iloc[:, 1:]\n",
    "table.iloc[:, 1:] = abundance / abundance.sum(axis=0)\n",
    "# rename the first column\n",
    "table.rename(columns={\"#OTU ID\": \"sequence\"}, inplace=True)\n",
    "# melt table\n",
    "melt_table = table.melt(id_vars=\"sequence\", value_name=\"abundance\", var_name=\"sample\")\n",
    "# Merging tooks several seconds\n",
    "feature_trait = pd.merge(melt_table, pred, how=\"inner\", on=\"sequence\")\n",
    "\n",
    "# Metadata from the paper\n",
    "metadata = pd.read_csv(\"metadata_SRP136143.csv\")\n",
    "metadata.drop(0, inplace=True)\n",
    "\n",
    "threshold_phylodistance = pd.read_csv(\"../../data/trait_autocorrelations/threshold_phylodistance.tsv\", sep= \"\\t\", index_col=0)\n",
    "threshold_phylodistance.rename(columns={\"cor_0.5\": \"threshold\"}, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff0bde1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Focus on genome size\n",
    "t = \"genome_size\"\n",
    "\n",
    "def rm_spurious_prediction(df, threshold_phylodistance, t):\n",
    "    threshold = threshold_phylodistance.loc[t, \"threshold\"]\n",
    "    # remove\n",
    "    if df[t+\"_nsti\"] > threshold:\n",
    "        return np.nan\n",
    "    else:\n",
    "        return df[\"abundance\"]\n",
    "\n",
    "feature_trait[\"fil_abd\"] = feature_trait.apply(\n",
    "    lambda x: rm_spurious_prediction(x, threshold_phylodistance, t), axis=1)\n",
    "\n",
    "feature_trait[t+\"_fil_abd\"] = feature_trait[\"fil_abd\"] * feature_trait[t]\n",
    "\n",
    "# Nomalization\n",
    "ratio = feature_trait.groupby(\"sample\", as_index=False)[\"fil_abd\"].sum()\n",
    "ratio[\"inv\"] = 1 / ratio[\"fil_abd\"]\n",
    "cwm = feature_trait.groupby(\"sample\", as_index=False)[t+\"_fil_abd\"].sum()\n",
    "cwm_normed = pd.merge(cwm, ratio, how=\"inner\", on=\"sample\")\n",
    "cwm_normed[t+\"_fil_abd_normed\"] = cwm_normed[t+\"_fil_abd\"] * cwm_normed[\"inv\"]\n",
    "\n",
    "cwm_nomred = pd.merge(cwm_normed, metadata, how=\"inner\", left_on=\"sample\", right_on=\"sample-id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a99a1e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(3, 2.5))\n",
    "\n",
    "# plot cwm along with salinity gradient\n",
    "x = cwm_nomred[\"SC\"].astype(float).values\n",
    "y = cwm_nomred[t+\"_fil_abd_normed\"].astype(float).values / 1e6\n",
    "ax.scatter(x, y, facecolor=\"None\", edgecolors=\"black\", s=30)\n",
    "\n",
    "# Fitting linear model\n",
    "lr = LinearRegression()\n",
    "lr.fit(x.reshape(-1, 1), y.reshape(-1, 1))\n",
    "ax.plot(x, lr.predict(x.reshape(-1, 1)), linestyle=\"dashed\", color=\"grey\")\n",
    "\n",
    "lr_sm = sm.OLS(y, sm.add_constant(x)).fit()\n",
    "conf_interval = lr_sm.conf_int(0.05)\n",
    "fill_x = np.sort(x)\n",
    "X = sm.add_constant(fill_x)\n",
    "y_lower = X.dot(conf_interval[:, 0])\n",
    "y_upper = X.dot(conf_interval[:, 1])\n",
    "ax.fill_between(fill_x, y_lower, y_upper, color=\"grey\", alpha=0.2)\n",
    "\n",
    "ax.set_title(\"Agricultural soil\", fontsize=14)\n",
    "ax.set_xlabel(r\"Salt content (g/kg)\", fontsize=14)\n",
    "ax.set_ylabel(\"Average genome size (Mb)\", fontsize=14)\n",
    "\n",
    "plt.savefig(\"../../results/08_application_soil_lake_microbiome/fig5b.pdf\", format=\"pdf\", dpi=300, facecolor=\"white\", bbox_inches=\"tight\", pad_inches=0.1)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bac2feature",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
